---
title: "GSE108565_WGCNA"
author: "Clodagh Murray"
date: "6/23/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# WGCNA 

```{r}
options(stringsAsFactors = FALSE)
library(dplyr)
library(genefilter)
library(WGCNA)
qnorm <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_qnorm.csv", row.names = 1)
datExpr <- as.matrix(qnorm)
datExpr <- varFilter(datExpr) #filter threshold top 30% of most variable genes 
dim(datExpr) #8590 genes
datExpr <- as.data.frame(t(datExpr)) #rows as samples, columns as genes
#write.csv(datExpr, file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_datExpr_WGCNA.csv', row.names = T)
## Step by step WGCN construction
gsg = goodSamplesGenes(datExpr, verbose = 3);
gsg$allOK
```

## Detect outliers

```{r}
sampleTree = hclust(dist(datExpr), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)
#pdf(file = "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)
# Plot a line to show the cut
abline(h = 115, col = "red")
```


# Make metadata with matching colnames
```{r}
samps = rownames(datExpr);
traitRows = match(samps, rownames(meta))
datTraits = meta[traitRows, ]
resistent <- c(1,1,1,1,1,1,1,0,0,0,0,0,0,0)
sensitive <- c(0,0,0,0,0,0,0,1,1,1,1,1,1,1)

# sample traits dataframe
datTraits <- data.frame(resistent, sensitive)

# Form a data frame analogous to expression data that will hold the clinical traits.
rownames(datTraits) = samps

# check phenotype data corresponds to exp data
table(rownames(datTraits) == rownames(datExpr))   
collectGarbage();
```
# Choosing the soft-thresholding power: analysis of network topology
```{r}
powers= c(c(1:10), seq(from =12, to=30, by=2)) #choosing a set of soft-thresholding powers
sft = pickSoftThreshold(datExpr, powerVector=powers, dataIsExpr = T, networkType = "signed hybrid", verbose =2, corFnc = "bicor") #call network topology analysis function
# Plot the results
sizeGrWindow(9,5)
par(mfrow= c(1,2))
cex1=0.9
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], xlab= "Soft Threshold (power)", ylab="Scale Free Topology Model Fit, signed R^2", type= "n", main= paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], labels=powers, cex=cex1, col="red")
abline(h=0.9, col="red")
plot(sft$fitIndices[,1], sft$fitIndices[,5], xlab= "Soft Threshold (power)", ylab="Mean Connectivity", type="n", main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1, col="red")
```
# Adjacency - check scale freeness
```{r}
softpower = 18
#plot adj in next section
adj= adjacency(datExpr,type = "signed hybrid", power = softpower, corFnc = "bicor")
head(adj[1:5,1:5])
# corFnc defaults to Pearson 
# When you have relatively few genes (<5000) use the following code
#k=as.vector(apply(adj,2,sum, na.rm=T))
# When you have a lot of genes use the following code
k=softConnectivity(datE=datExpr, power=softpower, type = "signed hybrid")
# Plot a histogram of k and a scale free topology plot
sizeGrWindow(10,5)
par(mfrow=c(1,2))
hist(k)
# Relatiely scale free - Poor fit to scale free topology may also indicate
scaleFreePlot(k, main="Check scale free topology\n")
```


# TOM from adjacency
To minimize effects of noise and spurious associations, we transform the adjacency into Topological Overlap Matrix,
and calculate the corresponding dissimilarity:
```{r}
# The TOM describes how well connected the genes are in respect of how many neighbors they share. 
# idea of TOM is to count the direct connection strengths as well as connection strengths “mediated” by shared neighbors
# signed network: preserves the sign of the relationship between the nodes while taking account the connections through shared neighbors in the same way the standard unsigned TOM does.
#dissTOM <- TOMdist(adj)
dissTOM <- read.csv('/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_dissTOM.txt', sep = "\t", row.names = 1)
#write.table(dissTOM, file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_dissTOM.txt', sep ='\t', col.names = T, row.names = T)
#take the distance of the TOM matrix
head(dissTOM[1:5,1:5])
```
# Clustering using TOM 
Module Definition; hierarchical clustering and tree cutting to define modules of co-expressed genes

```{r}
# Call the hierarchical clustering function from flashclust as faster than base hclust
geneTree = flashClust::hclust(as.dist(dissTOM), method = "average");
# average standard param; average dissimilarities between all objects
# each vertical line corresponds to a gene, clusters correspond to 'leafs' of genes
# dense branches indicate highly inter-connected, co-expressed genes
# module identification = identification of individual branches
# selecting the cut height leads to v different results, NB step
# use the dynamic tree cut package: bottom-up approach, merge branches evaluate (no of objects, gap between branches etc.),if meet criteria for being a module called seperate modules, otherwise merge
# can merge branches based on similiar eigengene values
sizeGrWindow(12,9)
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
labels = FALSE, hang = 0.04)
```

# Module decection
```{r}
minModuleSize = 30
# Module identification using dynamic tree cut:
dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,
deepSplit = 2, pamStage = F, pamRespectsDendro = FALSE,
minClusterSize = minModuleSize);
table(dynamicMods) # 43 modules detected
collectGarbage()
```
Plot dendogram, modules and traits
```{r}
# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)
# Convert traits to a color representation
traitColors = data.frame(labels2colors(datTraits))
dim(traitColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_gene_dend(deepsp = 2, PAM = T).pdf', wi = 9, he = 6)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05,
main = "Gene dendrogram and module colors") 
                    
#dev.off()
```

# Merging of modules whose expression profiles are very similar
```{r}
# Calculate eigengenes, match colours to dendogram
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
MEDissThres = 0.2#correlation of 80% to merge. 
sizeGrWindow(7, 6)
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_module_eigengene_clustering.pdf', wi = 9, he = 6)
plot(METree, main = "Clustering of module eigengenes",
xlab = "", sub = "")
abline(h=MEDissThres, col = "red")
#dev.off()
```

```{r}
# Call an automatic merging function; merge modules with eigengenes similarity >75%
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;
#plot shows the orignal module definition by dynamic tree cut and then the merging of modules via correlation in eigengene values
sizeGrWindow(12, 9)
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/Cluster_dend(pears22,eign_thresh=.2).pdf', wi = 9, he = 6)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
c("Dynamic Tree Cut", "Merged dynamic"),
dendroLabels = FALSE, hang = 0.03,
addGuide = TRUE, guideHang = 0.05)
#dev.off()
```
#Relating modules to external information and identifying important genes
# Gene Heatmap
Since we already have a summary profile (eigengene) for each module, we simply correlate eigengenes with external
traits and look for the most significant associations
```{r}
library(gplots)
#heatmap colours
myheatcol = colorpanel(250,'red',"orange",'lemonchiffon')
# For reproducibility, we set the random seed
set.seed(10);
#nSelect = 4000
select = sample(ncol(datExpr), size = 3000);
selectTOM = dissTOM[select, select];
# no simple way of restricting a clustering tree to a subset of genes, so we must re-cluster.
selectTree = hclust(as.dist(selectTOM), method = "average")
selectColors = mergedColors[select];
sizeGrWindow(9,9)
# Taking the dissimilarity to a power, makes the plot more informative by effectively changing the color palette; 
#setting the diagonal to NA also improves the clarity of the plot
plotDiss = selectTOM^9;
diag(plotDiss) = NA;
#takes long time to write out to pdf
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/gene_network_heatmap.pdf', wi = 9, he = 6)
TOMplot(plotDiss, selectTree, selectColors, main = "Network heatmap plot", col= myheatcol)
#dev.off()
```

```{r}
# Plot the dendrogram
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_eigengene_dendogram.pdf', wi = 9, he = 6)
plotEigengeneNetworks(mergedMEs, "Eigengene dendrogram", marDendro = c(0,4,2,0), marHeatmap = c(3,4,2,2),
plotHeatmaps = T)
#dev.off()
```

```{r}
# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs
table(moduleColors)
# Save module colors and labels for use in subsequent parts
save(MEs, moduleLabels, moduleColors, geneTree, file = "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_network_construction.RData")
```


# Module-Trait Relationships
```{r}
# Define numbers of genes and samples
nGenes = ncol(datExpr);
nSamples = nrow(datExpr);
# Recalculate MEs with color labels
# correlate with subtype 
MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, datTraits, use = "p"); #pearson correlation to assess correlation between merged eigengenes and traits
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);
# Form a data frame analogous to expression data that will hold the clinical traits.
samples = rownames(datExpr);
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_MTR_plot.pdf')
png("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_MTR.png")
#sizeGrWindow(10,6)
# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
          signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
xLabels = names(datTraits),
yLabels = names(MEs),
ySymbols = names(MEs),
colorLabels = TRUE,
colors = blueWhiteRed(50),
textMatrix = textMatrix,
setStdMargins = F,
cex.text = 0.5,
zlim = c(-1,1),
main = paste("Module-trait relationships"))
dev.off();
```

# Gene relationship to trait and important modules: Gene Significance and Module Membership
```{r}
#focus on modules associated with Resistence
resistent <- as.data.frame(datTraits$resistent);
names(resistent) = "Resistent"
# names (colors) of the modules
modNames = substring(names(MEs), 3)
# quantify associations of individual genes with our trait of interest by defining Gene Significance GS as the absolute value of the correlation between the gene and the trait
#For each module, we also define a quantitative measure of module membership MM as the correlation of the module eigengene and the gene expression profile. 
geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep=""); #p value for module membership of each gene
geneTraitSignificance = as.data.frame(cor(datExpr, resistent, use = "p")); #correlation of each gene  with trait
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples)); #the resultant p value
names(geneTraitSignificance) = paste("GS.", names(resistent), sep="");
names(GSPvalue) = paste("p.GS.", names(resistent), sep="");
```

# Correlating gene significance with clinical traits

```{r}
#average gene significance as a measure of module significance 
GS2=as.numeric(cor(resistent,datExpr, use="p"))
GeneSignificance=abs(GS2)
# Next module significance is defined as average gene significance.
ModuleSignificance=tapply(GeneSignificance, moduleColors, mean, na.rm=T)
sizeGrWindow(12,6)
png(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_gene_sig_barplot.pdf')
plotModuleSignificance(GeneSignificance, moduleColors) #could also use this to define pathway membership etc.
ModuleSignificance
#dev.off()
```

From the plot below, clearly, GS and MM are highly correlated, illustrating that genes highly significantly associated with a trait are often also the most important (central) elements of modules associated with the trait.
```{r}
MS_order <- sort(ModuleSignificance)
# calculate the module membership values (aka. module eigengene based connectivity kME)
datKME <- signedKME(datExpr, MEs )
datKME #ie how correlated, anticorrelated genes are with module eigengenes
#pdf(file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_GSvsMMplot.pdf')
par(mfrow = c(2, 2))
module = c("paleturquoise")
column = match(module, modNames);
moduleGenes = moduleColors==module
verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
abs(geneTraitSignificance[moduleGenes, 1]),
xlab = paste("Module Membership in", module, "module"),
ylab = "Gene significance for Resistence",
main = paste("Module membership vs. gene significance\n"),
cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
#par(mfrow = c(2, length(selectModules)/2))
dev.off()
```


# GSEA of Pale Turquoise Module
We have found modules with high association with our trait of interest, and have identified their central players by
the Module Membership measure. 
## Annotate probes - we have gene symbols but we want entrez IDs
```{r}
library(org.Hs.eg.db)
hs <- org.Hs.eg.db
#background set 
probes = rownames(qnorm)
probes2annot = AnnotationDbi::select(hs, 
       keys = probes,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
entrez <- match(probes, probes2annot$SYMBOL)
# the following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0.
```

```{r}
mod.paleturquoise <- names(datExpr)[moduleColors=="paleturquoise"] # 266 genes
names(mod.paleturquoise) <- rownames(mod.paleturquoise)
mod.paleturquoise_entrez <- AnnotationDbi::select(hs, 
       keys = mod.paleturquoise,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
#remove NAs
mod.paleturquoise_entrez <- mod.paleturquoise_entrez[rowSums(is.na(mod.paleturquoise_entrez)) != ncol(mod.paleturquoise_entrez), ]
```

```{r}
library(topGO)
library(Rgraphviz)
library(ggplot2)
library(enrichR)
#265 genes all up regulated
tT <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_tT.csv", row.names = 1)

list <- tT[mod.paleturquoise,]$adj.P.Val
names(list) <- rownames(tT[mod.paleturquoise,])

# selection is function that returns TRUE/FALSE for p-values<0.05
selection <- function(x) TRUE

allGO2genes <- annFUN.org(whichOnto="BP", feasibleGenes=NULL, mapping="org.Hs.eg.db", ID="symbol")

GOdata <- new("topGOdata", ontology="BP", allGenes=list, annot=annFUN.GO2genes, GO2genes=allGO2genes, geneSel=selection, nodeSize=10)
```
## Perform Enrichment 

```{r}
results.ks <- runTest(GOdata, algorithm="classic", statistic="ks")
```
# Turquoise Module Enrichment 
```{r}
goEnrichment <- GenTable(GOdata, KS=results.ks, orderBy="KS", topNodes=20)
goEnrichment$KS <- as.numeric(goEnrichment$KS)
goEnrichment <- goEnrichment[goEnrichment$KS<0.05,]
DT::datatable(goEnrichment, options = list(scrollX = TRUE, pageLength = 20, scroller = TRUE))
```

```{r}
#Plot the GO graph, color nodes by significance
showSigOfNodes(GOdata, score(results.ks), firstSigNodes = 5, useInfo = "all")
```


```{r}
library(rrvgo)
library(DOSE) #Disease ontology; description of gene product with disease perspectives
library(pathview)
library(clusterProfiler)
library(ggplot2)
#see what is negatively enriched in this
ego <- enrichGO(gene = mod.paleturquoise_entrez$ENTREZID, 
                universe = probes2annot$ENTREZID,
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "BP", #Biological Processes GO term
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_BP_summary <- data.frame(ego)
#View(cluster_BP_summary)
#png(file='/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturquoise_ORA_BP.png')
dotplot(ego, showCategory=10) + ggtitle("paleturquoise Module GO (Biological Process)")  +
  theme_classic()
#dev.off()

#write.csv(cluster_BP_summary, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturquoise_GO_BP_table")
```

## Revigo package to reduce redundancy in GO terms
```{r}
simMatrix <- calculateSimMatrix(cluster_BP_summary$ID,
                                orgdb="org.Hs.eg.db",
                                ont="BP",
                                method="Rel")

scores <- setNames(-log10(cluster_BP_summary$qvalue), cluster_BP_summary$ID)
reducedTerms <- reduceSimMatrix(simMatrix,
                                scores,
                                threshold=0.7,
                                orgdb="org.Hs.eg.db")
library(tidyverse)
reducedTerms <- reducedTerms %>% arrange(-score) %>% select(parentTerm, score, size)
table(reducedTerms$parentTerm)
heatmapPlot(simMatrix,
            reducedTerms,
            annotateParent=TRUE,
            annotationLabel="parentTerm",
            fontsize=6)
```

```{r}
png("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_GOterms_reduced.png")
scatterPlot(simMatrix, reducedTerms)
dev.off()

table(reducedTerms$parent)
```

```{r}
png("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_GOterms_treemap.png")
treemapPlot(reducedTerms)
dev.off()
```



```{r}
ego <- enrichGO(gene = mod.paleturquoise_entrez$ENTREZID, 
                universe = probes2annot$ENTREZID,
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "CC", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturquoise_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("PaleTurqoise Module GO (Cellular Component)")  +
  theme_classic()
#dev.off()
```

```{r}
ego <- enrichGO(gene = mod.paleturquoise_entrez$ENTREZID, 
                universe = probes2annot$ENTREZID,
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_BP_summary <- data.frame(ego)
#View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturquoise_ORA_MF.pdf')
dotplot(ego, showCategory=10) + ggtitle("PaleTurqoise Module GO (Molecular Function)")  +
  theme_classic()
#dev.off()
```

```{r}
library(stringr)
library(ggplot2)
#KEGG Pathway Enrichment Analysis
ekegg <- enrichKEGG(gene = mod.paleturquoise_entrez$ENTREZID, 
                organism = "hsa",
                keyType = 'kegg',
           #     keyType = "ENTREZID",
                qvalueCutoff = 0.05, 
                pAdjustMethod = "BH", 
                universe = probes2annot$ENTREZID)
cluster_KEGG_summary <- data.frame(ekegg)
write.csv(cluster_KEGG_summary, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturqoise_KEGG.csv")
#View(cluster_KEGG_summary)
#KeGG pathway looks very noisey
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_paleturqoise_ORA_KEGG.pdf')
dotplot(ekegg, showCategory=10, font.size = 8) + scale_y_discrete(labels = function(x) str_wrap(x, width = 20)) +
  ggtitle("paleturquoise Module GO (Molecular Function)")  + 
  theme_classic() 
#dev.off()
```

Overlap between DEGs and paleturquoise module
```{r}
library(AnnotationDbi)
sig_exprs <- read.csv("GSE108565_DEG_counts.csv", sep = "\t", row.names = 1) #617 
# get sig genes into same format as module genes
DEGs <- AnnotationDbi::select(hs, 
       keys = rownames(sig_exprs),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
DEGs <- DEGs[-204,] #remove NA
```

# Venn Diagrsm of overlap 
```{r}
#Use the Venn Diagram package in R
library(VennDiagram)
colours <- c('#FFC0CB', '#40E0D0')
venn.diagram(x = list(DEGs$ENTREZID, mod.paleturquoise_entrez$ENTREZID),
             category.names = c('DEG\'s', 'Paleturquoise Module'),
           filename = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_DGE_0.5_paleturquoise_Module_Overlap.png',
             output = T,
             imagetype = 'png',
             scaled = F,
             col = 'grey70',
             fill = colours,
             cat.cex = 1,
             margin = 0.15 )
#display image
options(repr.plot.height=12, repr.plot.width = 12)
library('png')
pp <- readPNG("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_DGE_0.5_paleturquoise_Module_Overlap.png")
plot.new()
rasterImage(pp, 0,0,1.1, 1.1)
```

Also look at enrichment for up and downregualted genes, assess if enrichment is significant
#focus on turquoise and pink module for down and up regulated genes

```{r}
#upregulated and downregulated genes from DEG analysis with DESeq2
#using rownames 
tT <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_tT.txt", sep = ",", row.names = 1)

# convert gene symbols to entrez id
up_reg <- tT[tT$adj.P.Val <= 0.05 & tT$logFC >= 0.5, ];
up_reg<- AnnotationDbi::select(hs, 
       keys = rownames(up_reg),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
up_reg <- up_reg[-328,] #remove NA

# convert gene symbols to entrez id
down_reg <- tT[tT$adj.P.Val <= 0.05 & tT$logFC < -0.5, ];
down_reg <- AnnotationDbi::select(hs, 
       keys = rownames(down_reg),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
#check if significant overlap between these two with the pink and turquoise modules
#use phypher function to test if overlap is significant 
#upreg
#182 upregulated genes in the paleturquoise module
length(intersect(up_reg$ENTREZID, mod.paleturquoise_entrez$ENTREZID))
# 0 downregulated genes in the paleturquoise module
length(intersect(down_reg$ENTREZID, mod.paleturquoise_entrez$ENTREZID))
#downreg
```

```{r}
#test for the statistical significance of these overlaps using hypergeometric test
Overlap <- length(intersect(up_reg$ENTREZID, mod.paleturquoise_entrez$ENTREZID))
Group1 <- length(up_reg$ENTREZID)
Group2 <- length(mod.paleturquoise_entrez$ENTREZID)
Background <- length(probes) #all genes used in the wgcna analysis (~8.5k genes)
pval.modpaleturquoise <- phyper(Overlap,Group2, Background - Group2, Group1, lower.tail = F )  #if less than 0.05 assume significant enrichment
pval.modpaleturquoise # 2.188301e-285 VERY significantly enriched for up regulated genes
#FF is pink, #40E is turquoise
#colours <- c('#CE2F17', '#E0FFFF', '#88D969', '#FFC0CB', '#40E0D0')
colours <- c('#FFC0CB', '#40E0D0')
venn.diagram(x = list(mod.paleturquoise_entrez$ENTREZID, up_reg$ENTREZID),
             category.names = c('Paleturquoise Module','Upregulated DEG\'s'),
             filename = '/home/clodagh/MASTERS PROJECT/GSE37614/Upregpaleturquoise_Module_0.5_Overlap.png',
             output = T,
             main = '',
             sub = 'Overlap of DEG\'s and Paleturqoise Module',
             imagetype = 'png',
             scaled = F,
             col = 'grey70',
             alpha = rep(0.5, 2),
             fill = colours,
             cat.cex = 1,
             fontfamily = "sans",
            sub.pos = c(.5,.80),
            cat.default.pos = "outer",
            hyper.test = T,
            total.population = Background,
            lower.tail = F,
            cat.pos = c(3, 0),
             margin = 0.15 )
options(repr.plot.height=12, repr.plot.width = 12)
library('png')
pp <- readPNG('/home/clodagh/MASTERS PROJECT/GSE37614/Upregpaleturquoise_Module_0.5_Overlap.png')
plot.new()
rasterImage(pp, 0,0,1.1, 1.1)
```

```{r}
#use these to filter the modules filter paleturquoise module for export to cytoscape
filter_paleturquoise <- intersect(DEGs$SYMBOL, mod.paleturquoise_entrez$SYMBOL)
which(DEGs$SYMBOL == "IGFBP10")
tT["CDKN2B",]
```
# Export Network to Cytoscape

From tutorial
```{r}
# Recalculate topological overlap if needed
TOM = TOMsimilarityFromExpr(datExpr, power = 18, corType = "bicor", networkType = "signed hybrid");
head(TOM[1:5,1:5])

```


#Importing paleturquoise module

```{r}
library(WGCNA)
module = "paleturquoise"
# Read in the annotation file
annot <- AnnotationDbi::select(hs, 
       keys = colnames(datExpr),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
#remove duplicates
annot <- distinct(annot, SYMBOL, .keep_all = T)
# Select modules
inModule= is.finite(match(annot$SYMBOL, filter_paleturquoise)) #so this now filters the module by intersect
#inModule
#inModule = is.finite(match(moduleColors, modules));
modProbes = annot[inModule,]$SYMBOL;
modGenes = annot[inModule,]$ENTREZID;
# Select the corresponding Topological Overlap
modTOM = TOM[inModule, inModule];
dimnames(modTOM) = list(modGenes, modGenes)
# Export the network into edge and node list files Cytoscape can read
dim(modTOM)
cyt = exportNetworkToCytoscape(modTOM,
edgeFile = paste("CytoscapeInput-edges-", paste(module, collapse="-"), "TOM0.25.txt", sep=""),
nodeFile = paste("CytoscapeInput-nodes-", paste(module, collapse="-"), "TOM0.25.txt", sep=""),
threshold = 0.2,
nodeNames = modProbes,
altNodeNames = modGenes,
nodeAttr = moduleColors[inModule]);

weight <- head(cyt$edgeData[order(-cyt$edgeData$weight),], 0.1*nrow(cyt$edgeData))
write.csv(weight, "/home/clodagh/MASTERS PROJECT/GSE37614/pale_turquoise_edgedata_sortedbyweight.csv", row.names =T)
```


```{r}
library(igraph)
edges <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/CytoscapeInput-edges-paleturquoise.txt", sep = "\t")
edges <- as.matrix(edges[c(1:30),c(5,6)])

igraph <- plot(graph_from_edgelist(edges))
```

```{r}
#write out node and edge table, wil use these  to construct network
write.table(cyt$edgeData, file = "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_cyt_paleturquoiseedge_noMMGS_noTOMfilter.txt", sep = "\t ", row.names = F, col.names = TRUE, quote = F)
write.table(cyt$nodeData, file = "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_cyt_paleturquoisenode_noMMGS_noTOMfilter.txt", sep = "\t ", row.names = F, col.names = TRUE, quote = F)
```

## Cytoscape Export 
```{r}
installApp('stringApp')
library(RCy3)
cytoscapePing()
edge <- read.delim("/home/clodagh/MASTERS PROJECT/GSE37614/CytoscapeInput-edges-paleturquoise.txt")
colnames(edge) <- c("source", "target", "weight", "direction", "fromAltName", "toAltName")
node <-  read.delim("/home/clodagh/MASTERS PROJECT/GSE37614/CytoscapeInput-nodes-paleturquoise.txt")
colnames(node) <- c("id", "altName", "node_attributes")

createNetworkFromDataFrames(node, edge, title = "First 50")
```



# Centrality
```{r}
central <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/CytoscapeInput-edges-paleturquoise-network_analyzer.csv")
rownames(central) <- central$name

top_central <- head(central[order(-central$BetweennessCentrality),],.10*nrow(central))

central_genes <- rownames(top_central)
```



Network of all DEGs

```{r}
library(WGCNA)
DEGs <- read.csv("GSE108565_DEG_list.csv", sep = ",")
DT::datatable(DEGs, options = list(scrollX = TRUE, pageLength = Inf, scroller = TRUE))
DEG_list  <- DEGs$ID
interferon_ti <- c("OAS3","C19orf66", "OASL", "SAMHD1","ISG20","STAT2","MX2","OAS2", "IFI6","IFNAR2", "XAF1", "IFIT3","IFIT2", "PSMB8","ISG15", "RSAD2", "IFI35")
interferon_a_b <- c("OAS3","DHX58","OASL","SP110","RTP4","SAMHD1","PARP12","IFIH1","UBE2L6","DTX3L","ISG20","EPSTI1","TRIM25","SAMD9L","MX2","OAS2","IFI6","PARP9","XAF1","GBP5","IFI44","IFI44L","IFIT3","IFIT2","ISG15","RSAD2","DDX60","IFI35","PARP14","RNF213")

int = is.finite(match(annot$SYMBOL, interferon_a_b)) #so this now filters the module by intersect
#inModule
#inModule = is.finite(match(moduleColors, modules));
modProbes = annot[int,]$SYMBOL;
modGenes = annot[int,]$ENTREZID;
# Select the corresponding Topological Overlap
modTOM <- TOM[int, int];
dimnames(modTOM) = list(modGenes, modGenes)
# Export the network into edge and node list files Cytoscape can read
cyt_int = exportNetworkToCytoscape(modTOM,
edgeFile = paste("Cytoscape-edges-", paste("Int", collapse="-"), ".txt", sep=""),
nodeFile = paste("Cytoscape-nodes-", paste("Int", collapse="-"), ".txt", sep=""),
threshold = 0.1,
nodeNames = modProbes,
altNodeNames = modGenes,
nodeAttr = moduleColors[int]);
```



## Construct network of interferon alpha signalling and colour by logFC to see if enriched and co-expressed
```{r}
library(stringr)
string <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/pale_turquoise_STRING_clusters.csv")
geen <- c("OAS3","DHX58","OASL","SP110","RTP4","PARP12","IFIH1","UBE2L6","DTX3L","ISG20","EPSTI1","TRIM25","SAMD9L","MX2","OAS2","IFI6","PARP9","XAF1","IFI44","IFI44L","IFIT3","IFIT2","ISG15","RSAD2","DDX60","IFI35","PARP14","RNF213")

# extract TOM for above genes and construct network
DE = is.finite(match(annot$SYMBOL, geen)) 
DE <- which(DE)
#so this now filters the module by intersect
#inModule
#inModule = is.finite(match(moduleColors, modules));
modProbes = annot[DE,]$SYMBOL;
modGenes = annot[DE,]$ENTREZID;
# Select the corresponding Topological Overlap
modTOM = TOM[DE, DE];
dim(modTOM)
dimnames(modTOM) = list(modProbes, modProbes)
# Export the network into edge and node list files Cytoscape can read
cyt = exportNetworkToCytoscape(modTOM,
edgeFile = paste("Cytoscape-edges-", paste("Inf_alpha", collapse="-"), ".txt", sep=""),
nodeFile = paste("Cytoscape-nodes-", paste("Inf_alpha", collapse="-"), ".txt", sep=""),
threshold = 0.01,
nodeNames = modProbes,
altNodeNames = modGenes,
nodeAttr = moduleColors[DE]);
```


# Import analysed network and info on top 20 hub genes
```{r}
hub <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/pale_turquoise_degree_infocsv.csv")
hub <- hub[, c("name", "degree.layout", "logFC" )]
hub <- hub[order(-hub[,3]),]
hub <- hub[1:20,]
colnames(hub) <- c("Gene", "Degree", "Log2FC")
write.csv(hub, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_hubtop20.csv", row.names = F)
```

# Heatmap of hub genes
```{r}
library(pheatmap)
labs <- c(rep("resistent", 7), rep("sensitive", 7))
hub_exprs <- qnorm[rownames(qnorm) %in% hub$Gene, ]
pheatmap(hub_exprs, trace="none", scale="row", cexRow=0.2, cexCol=0.7, annotation_col = annotation_for_heatmap, show_rownames = T, cluster_cols = T, labels_col = labs, main = "Heatmap of top 20 hub genes")
#dev.off()
```




# Survival Analysis
```{r}
library(TCGAbiolinks)

GDCprojects <- getGDCprojects()
head(GDCprojects[c("project_id","name")])

TCGAbiolinks::getProjectSummary("TCGA-BRCA")

query_TCGA = GDCquery(project = "TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      experimental.strategy = "RNA-Seq",
                      workflow.type = "STAR - Counts",
                      sample.type = c("Primary Tumor", "Solid Tissue Normal"))

brca_res <- getResults(query_TCGA)
colnames(brca_res)
summary(brca_res$sample_type)
brca_res$sample_type <- as.factor(brca_res$sample_type)
summary(brca_res$data_type)



library(TCGAbiolinks)
# Survival Analysis SA

clinical_patient_Cancer <- GDCquery_clinic("TCGA-BRCA","clinical")
dataBRCAcomplete <- log2(BRCA_rnaseqv2)

tokenStop <- 1

tabSurvKMcomplete <- NULL

for( i in 1: round(nrow(dataBRCAcomplete)/100)){
    message( paste( i, "of ", round(nrow(dataBRCAcomplete)/100)))
    tokenStart <- tokenStop
    tokenStop <- 100 * i
    tabSurvKM <- TCGAanalyze_SurvivalKM(
        clinical_patient_Cancer,
        dataBRCAcomplete,
        Genelist = rownames(dataBRCAcomplete)[tokenStart:tokenStop],
        Survresult = F,
        ThreshTop = 0.67,
        ThreshDown = 0.33
    )
    
    tabSurvKMcomplete <- rbind(tabSurvKMcomplete,tabSurvKM)
}

tabSurvKMcomplete <- tabSurvKMcomplete[tabSurvKMcomplete$pvalue < 0.01,]
tabSurvKMcomplete <- tabSurvKMcomplete[order(tabSurvKMcomplete$pvalue, decreasing=F),]

tabSurvKMcompleteDEGs <- tabSurvKMcomplete[
    rownames(tabSurvKMcomplete) %in% dataDEGsFiltLevel$mRNA,
]
```



