---
title: "GSE37614 batch correction"
author: "Clodagh Murray"
date: "5/19/2022"
output: html_document
---


### Libraries
```{r, message =F}
library(oligo)
library(Biobase)
library(ggpubr)
library(data.table)
library(DT)
library(dplyr)
library(ggpubr)
library(ggplot2)
library(limma)
library(RColorBrewer)
library(pheatmap)
library(PCAtools)
```

## Raw Data Staging
### 4 samples were analysed in both batches 1 and 2, creating 4 extra samples in the raw data - update metadata to match raw data 

```{r}
SDRF <- read.delim("/home/clodagh/MASTERS PROJECT/GSE37614/E-GEOD-37614.sdrf_fixed.txt")
rownames(SDRF) <- SDRF$Array.Data.File
DT::datatable(SDRF, options = list(scrollX = TRUE, pageLength = 23, scroller = TRUE))
```

## Count matrix matching metadata?
```{r}
## same as first attempt tbh
gen_raw <- function(path, pattern){
  files = list.files(path, pattern, full.names=T, recursive=T, include.dirs=T)
  mat = as.data.frame(do.call(cbind, lapply(files, function(x) data.table::fread(x, stringsAsFactors=FALSE))))
  ID_REF = mat[,1]
  mat = as.data.frame(mat[,!grepl("ID_REF", colnames(mat))])
  rownames(mat) = ID_REF
  return(mat)
}

raw.dat <- gen_raw("/home/clodagh/MASTERS PROJECT/GSE37614/raw/", "\\.txt$")

## attach sample names to cols
file_names <- list.files("/home/clodagh/MASTERS PROJECT/GSE37614/raw/", "\\.txt$", full.names=T, recursive=T, include.dirs=T)
file_names <- sub(".*/", "", file_names)
sample_names <- sub("\\_.*", "", file_names)
length(sample_names) == ncol(raw.dat)
```

## Append IDs to matrix

```{r}
#  append batch ID's to the sample names for unique colnames. 
SDRF$samp_names <- paste(SDRF$Assay.Name, SDRF$Characteristics..batch., sep=".")
SDRF$samp_names

```
```{r}
# make sure columns match file contents
SDRF <- SDRF[match(file_names, SDRF$Array.Data.File),]
colnames(raw.dat) <- SDRF$samp_names
samples <- SDRF

## the below is correct 
head(raw.dat[1:5, 1:5])
```

## Make Expression Set manually

```{r}
SDRF <- Biobase::AnnotatedDataFrame(SDRF)
rownames(SDRF) <- colnames(raw.dat)
my_set <- Biobase::ExpressionSet(assayData = as.matrix(raw.dat), phenoData = SDRF)
my_set
```

## Raw Data Boxplot
```{r}
SDRF@data$FactorValue..PATIENT.ID.
SDRF@data$Characteristics..class.[23] <- "Her2+"
sml <- as.character(as.numeric(as.factor(SDRF@data$Characteristics..class.))) 
gs <- factor(sml)
groups <- SDRF@data$Characteristics..class. 
orientation <- unique(groups)
orientation <- c(orientation[2], orientation[3], orientation[1])
levels(gs) <- orientation
my_set@phenoData$group <- gs
ord <- order(gs)

palette <- palette(c("#1B9E77", "#7570B3", "pink"))

par(mar=c(7,4,2,1))
title <- "GSE37614 Log2 Raw Signals"
boxplot(log2(exprs(my_set[,ord])), boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", orientation, fill=palette(), bty="n")
```


## Quantile normalization

```{r}
quantile_normalisation <- function(df){
  df_rank <- apply(df,2,rank,ties.method="min")
  df_sorted <- data.frame(apply(df, 2, sort))
  df_mean <- apply(df_sorted, 1, mean)

  index_to_mean <- function(my_index, my_mean){
    return(my_mean[my_index])
  }
df_final <- apply(df_rank, 2, index_to_mean, my_mean=df_mean)
  rownames(df_final) <- rownames(df)
  return(df_final)
}

qnorm <- quantile_normalisation(raw.dat)
```


## Update ExpressionSet Object 

### Boxplot of Qnorm

```{r}
exprs(my_set) <- qnorm

## boxplot
par(mar=c(7,4,2,1))
title <- "GSE37614 Log2 Quantile Normalization"
boxplot(log2(qnorm[,ord]), boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", orientation, fill=palette(), bty="n")
```
## Expression Distribution

```{r}
ex <- exprs(my_set)
par(mar=c(4,4,2,1))
title <- paste ("Normalized Expression Density", sep ="")
limma::plotDensities(log2(ex), group=gs, main=title, legend ="topright")
```

## Probe Filtering

```{r}
# Probe filtering
medians <- rowMedians(ex)

hist(log2(medians + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
```


## Remove Multimapping probes

```{r, message = F}
probes <- rownames(my_set)

library(illuminaHumanv4.db)
```

```{r}
annotation <- AnnotationDbi::select(illuminaHumanv4.db,
                                    keys = probes,
                                    columns = c("SYMBOL", "GENENAME", "ENTREZID"),
                                    keytype = "PROBEID")
```

```{r}
## 'select()' returned 1:many mapping between keys and columns
annotation <- subset(annotation, !is.na(ENTREZID))

## resolve multi maps. 

ann_grouped <- group_by(annotation, PROBEID)
ann_sum <- dplyr::summarize(ann_grouped, no_of_matches = n_distinct(ENTREZID))

ann_flt <- dplyr::filter(ann_sum, no_of_matches > 1)

remove_id <- (probes %in% ann_flt$PROBEID)

table(remove_id)
```
### Too many genes
*manual filtering*

```{r}
annotations <- read.table("/home/clodagh/MASTERS PROJECT/GSE37614/A-GEOD-10558.adf.txt", header=T, sep="\t")
mart <- biomaRt::useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
attr <- biomaRt::listAttributes(mart)
genbank_map <- biomaRt::getBM(attributes = c("refseq_mrna", "hgnc_symbol", "gene_biotype"), mart = mart, useCache = FALSE)
genbank_map <- genbank_map[!(is.na(genbank_map$refseq_mrna) | genbank_map$refseq_mrna==""),]
genbank_map <- subset(genbank_map, genbank_map$gene_biotype == "protein_coding")

# pop off the version ID
annotations$genbank <- annotations$Reporter.Database.Entry..genbank.

annotations$genbank <- gsub("\\.\\d+$", "", annotations$genbank)
# merge
master <- merge(genbank_map, annotations, by.x="refseq_mrna", by.y="genbank")
```

```{r}
master_sub <- master[,c(4,2)]
master_sub <- master_sub[!(is.na(master_sub$hgnc_symbol) | master_sub$hgnc_symbol==""),]

# remove multi mapping probes
master_sub <- master_sub %>% distinct(hgnc_symbol, .keep_all = T)
dim(master_sub)

```

```{r}
qnorm <- subset(qnorm, rownames(qnorm) %in% master_sub$Reporter.Name)
dim(qnorm)
```

```{r}
master_sub <- master_sub[match(rownames(qnorm), master_sub$Reporter.Name),]
rownames(qnorm) <- master_sub$hgnc_symbol

# qnorm now annotated with gene symbol instead of reporter name
head(qnorm[1:3,1:3])

```



```{r}
medians <- rowMedians(qnorm)

hist(log2(medians + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
```

```{r}
medians <- log2(rowMedians(qnorm) + 1)
qnorm <- as.data.frame(qnorm)
qnorm$medians <- medians

qnorm <- subset(qnorm, qnorm$medians > 7.9)
qnorm <- as.matrix(qnorm[,1:(ncol(qnorm)-1)]) # get rid of last col (medians) 

# plot again
hist(log2(qnorm + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
dim(qnorm)
#write.csv(qnorm, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_qnorm.csv", row.names = T)
qnorm <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_qnorm.csv", row.names = 1)

```


```{r}
qnorm <- qnorm[,-23]
meta <- my_set@phenoData@data
meta <- meta[-23,] #remove outlier
meta <- meta[,c(4, 5)]
colnames(meta) <- c("Batch", "Cancer_Subtype")
#write.table(meta, "metadata.csv", row.names=T, sep="\t")
#write.csv(qnorm, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_qnorm.csv", row.names=T)
```

-----------------------------------------------------------------------
#### BEGIN HERE #############
```{r}
qnorm <- read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_qnorm.csv", row.names=1)
meta <-  read.csv("/home/clodagh/MASTERS PROJECT/GSE37614/metadata.csv", sep = "\t")
```


# Exploratory Data Analysis
## Sample heatmap

```{r}
phenotype_names <- meta$Cancer_Subtype
batch <- meta$Batch
annotation_for_heatmap <- 
  data.frame(subtype = phenotype_names,  Batch= batch)
row.names(annotation_for_heatmap) <- rownames(meta)
```



```{r}
dists <- as.matrix(dist(t(qnorm), method = "manhattan"))

rownames(dists) <- rownames(meta)
hmcol <- colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255)
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(
  subtype= c("TNBC" = "chartreuse4", "Her2+" = "burlywood3", "ER+" = "purple"),
  Batch = c(b1= "blue4", b2 = "cadetblue2")
                   )
pheatmap(dists, col = (hmcol), cluster_rows= T,
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")
```



### PCA Cancer Subtypes
*2 groups - largely driven by batch*

```{r}
PCA <- prcomp(t(log2(qnorm)), scale = TRUE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Subtype = meta$Cancer_Subtype,
                    Batch = meta$Batch)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Subtype", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed quantile normalized expression data",
                  subtitle = "Cancer Subtypes",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = FALSE, star.plot = F,
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

### PCA Batch Effects
```{r}
dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Batch = meta$Batch)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Batch", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed quantile normalized expression data",
                  subtitle = "Batch Effects (Batch 1, Batch 2)",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = TRUE, star.plot = TRUE,
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```


## Pearsons Correlation btween PCs and batch and cancer

```{r}
pca <- PCAtools::pca(log2(qnorm + 1), metadata = meta)

PCAtools::eigencorplot(pca, components = PCAtools::getComponents(pca, 1:10),
    metavars = c('Batch','Cancer_Subtype'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2, fontCorval = 2, posLab = 'all', rotLabX = 45,
    scale = TRUE,
    main = bquote(PC ~ Pearson ~ r^2 ~ clinical ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
```


## Limma Differential Expression with batch covariate

```{r}
groups <- as.factor(meta$Cancer_Subtype)
batch <- as.factor(meta$Batch)
design <- model.matrix(~0 + groups + batch)
gps <- c("ER", "HER", "TNBC", "b2")
colnames(design) <- gps
#rm_batch <- limma::removeBatchEffect(log2(qnorm + 1), batch = meta$Batch, design=design)
aw <- arrayWeights(log2(qnorm)+1, design)

fit <- limma::lmFit(log2(qnorm)+1, design, weights = aw)
#write.table(rm_batch, file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_rm_batch.txt', sep ='\t', col.names = T, row.names = T)

# Set up contrasts of interest
cts <- c("HER-ER", "HER-TNBC", "ER-TNBC")
cont.matrix <- limma::makeContrasts(contrasts=cts, levels=design)

fit2 <- limma::contrasts.fit(fit, cont.matrix)
fit2 <- limma::eBayes(fit2,0.01)
res <- limma::decideTests(fit2)
summary(res)
```

## PCA Cancer Subtype (Corrected)

```{r}
PCA <- prcomp(t(log2(qnorm)), scale = TRUE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Subtype = meta$Cancer_Subtype)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Subtype", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed batch corrected data",
                  subtitle = "Cancer Subtypes",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = F, star.plot = T, 
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

## PCA Batch Effect (Corrected)

```{r}
dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Batch = meta$Batch)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Batch", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed batch corrected data",
                  subtitle = "Batch Effects: Batch1, Batch2",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = T, star.plot = T, 
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

## Sample heatmap (batch correcetd)


```{r}
dists <- as.matrix(dist(t(log2(qnorm)), method = "manhattan"))

rownames(dists) <- rownames(meta)
hmcol <- colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255)
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(
  subtype= c("TNBC" = "chartreuse4", "Her2+" = "burlywood3", "ER+" = "purple"),
  Batch = c(b1= "blue4", b2 = "cadetblue2"))

# Remove outlier? 
pheatmap(dists, col = (hmcol),
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE,
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "GSE37614 Clustering heatmap normalized samples")
```


# Differentially Expressed Genes
## No coefficient
```{r}
# all genes with a p value < 0.05 and abs(logFC) over 0.5 for any contrast
tT <- topTable(fit2, adjust="fdr", number=Inf)
#write.csv(tT, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_tT.csv", row.names =T)
tT["RHOA",]
top_sig <- tT[tT$adj.P.Val <= 0.05 & (abs(tT[,1]) > 0.5 | abs(tT[,2]) > 0.5), ];
nrow(top_sig)                  
```
## Heatmap 
```{r}
top_sigg <- qnorm[rownames(qnorm) %in% rownames(top_sig),]
top_sigg<- t(top_sigg)
top_sigg <- scale(top_sigg, center=T, scale = T)
top_sigg <- t(top_sigg)
dim(top_sigg)
# logFC >= 0.5 & adjusted p-value <= 0.05
#write.table(top_sigg, "GSE37614_DEGs_overall.csv", row.names=T, sep="\t")

annot_col <- data.frame(row.names = colnames(qnorm), Group = factor(meta$Cancer_Subtype))

cols <- c("pink", "blue", "orange")
names(cols) <- c("ER+", "Her2+", "TNBC")
annot_colors <- list(Group = cols)

library(viridis)
```


```{r}
ord <- c(11,12,13,14,15,16,22,1,2,3,4,17,18,19,5,6,7,8,9,10,20,21)
pheatmap::pheatmap(top_sigg[,ord], col =  colorRampPalette(rev(brewer.pal(n = 10, name =
"RdYlBu")))(100),cluster_cols =  F, cluster_rows = T, show_rownames=F, show_colnames = F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "GSE37614 388 DEGS (logFC > 0.5 & adj.p.val < 0.05) ")
```



### coef = 1 (ER-Her2)
```{r}
#Create the top table for the comparison (coef) we want
top1 <- topTable(fit2, coef=1, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg1 <- top1[top1$adj.P.Val <= 0.05 & abs(top1$logFC) >= 0.5, ];
nrow(up_reg1)
```
## Volcano plot of ER-HER2
```{r}
plot(top1$logFC, -log10(top1$adj.P.Val), pch="*", xlab="Log2 Fold Change", ylab="-10log (adjusted p-value)", main = "ER vs. Her2 DEGs")
abline(h=-log10(0.05), v=c(-0.5, 0.5), col="red", lty=2)
points(up_reg1$logFC, -log10(up_reg1$adj.P.Val), col="red", pch="*")
```

```{r}
library(gplots)
sig_HER_ER <- qnorm[rownames(qnorm) %in% rownames(up_reg1),]
sig_HER_ER  <- t(sig_HER_ER)
sig_HER_ER  <- scale(sig_HER_ER , center=T, scale = T)
sig_HER_ER  <- t(sig_HER_ER)
dim(sig_HER_ER)
pheatmap::pheatmap(sig_HER_ER[,ord], col =  colorRampPalette(rev(brewer.pal(n = 10, name =
"RdYlBu")))(100),cluster_cols =  F, cluster_rows = T, show_rownames=F, show_colnames = F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "HER2 vs ER+ DEGs")
#heatmap.2(sig_HER_ER, trace="none", scale="row", col="redgreen", cexRow=0.2, cexCol=0.7, main = "HER vs ER DEGS")
```





### coef = 2 (Her2-TNBC)
```{r}
top2 <- topTable(fit2, coef=2, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg2 <- top2[top2$adj.P.Val <= 0.05 & abs(top2$logFC) >= 0.5, ];
nrow(up_reg2)
```
## Volcano plot of HER2 vs TNBC
```{r}
plot(top2$logFC, -log10(top2$adj.P.Val), pch="*", xlab="Log2 Fold Change", ylab="-10log (adjusted p-value)", main = "Her2 vs TNBC DEGs")
abline(h=-log10(0.05), v=c(-0.5, 0.5), col="red", lty=2)
points(up_reg2$logFC, -log10(up_reg2$adj.P.Val), col="red", pch="*")
```

```{r}
library(gplots)
sig_HER_TNBC <- qnorm[rownames(qnorm) %in% rownames(up_reg2),]
sig_HER_TNBC  <- t(sig_HER_TNBC)
sig_HER_TNBC  <- scale(sig_HER_TNBC , center=T, scale = T)
sig_HER_TNBC  <- t(sig_HER_TNBC)

pheatmap::pheatmap(sig_HER_TNBC[,ord], col =  colorRampPalette(rev(brewer.pal(n = 10, name =
"RdYlBu")))(100),cluster_cols =  F, cluster_rows = T, show_rownames=F, show_colnames = F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "HER2 vs TNBC DEGs")
#heatmap.2(sig_HER_TNBC, trace="none", scale="row", col="redgreen", cexRow=0.2, cexCol=0.7, main = "HER vs TNBC DEGS")
```


### coef = 3 (TNBC-ER)
```{r}
top3 <- topTable(fit2, coef=3, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg3 <-  top3[top3$adj.P.Val <= 0.05, ];
nrow(up_reg3)
```

```{r}
library("VennDiagram")
genes <- list(c1= rownames(sig_HER_ER),
             c2=rownames(sig_HER_TNBC),
             c3=rownames(up_reg3))

ol = calculate.overlap(x = genes)
ol_size=sapply(ol, length)

# move to new plotting page
grid.newpage()
  
Background <- nrow(qnorm)
colours <- c('#CE2F17', '#E0FFFF', '#88D969')
# create Venn diagram with three sets
venn.diagram(x = list(rownames(sig_HER_ER),
             rownames(sig_HER_TNBC),
             rownames(up_reg3)),
             category.names = c('HER-ER', 'HER-TNBC', 'ER-TNBC'),
             filename = '/home/clodagh/MASTERS PROJECT/GSE37614/DEG_Overlap_Ven.png',
             output = T,
             main = 'GSE37614 DEGs (|logFC| > 0.5 and adj.p.val <0.05)',
             imagetype = 'png',
             scaled = F,
             col = 'grey70',
             fill = colours,
             cat.cex = 1,
             fontfamily = "sans",
            sub.pos = c(.5,.80),
            cat.default.pos = "outer",
            hyper.test = T,
            total.population = nrow(qnorm),
            lower.tail = F,
             margin = 0.15 )
options(repr.plot.height=12, repr.plot.width = 12)
library('png')
pp <- readPNG('/home/clodagh/MASTERS PROJECT/GSE37614/DEG_Overlap_Ven.png')
plot.new()
rasterImage(pp, 0,0,1.1, 1.1)
```


## Heatmap 

```{r}
DEG_names <- c(rownames(up_reg1), rownames(up_reg2))
DEG_names <- unique(DEG_names)
e <- qnorm
sig_exprs <- e[rownames(e) %in% DEG_names,]
sig_exprs <- t(sig_exprs)
sig_exprs <- scale(sig_exprs, center=T, scale = T)
sig_exprs <- t(sig_exprs)
dim(sig_exprs)
annot_col <- data.frame(row.names = colnames(qnorm), Group = meta$Cancer_Subtype)

cols <- c("pink", "blue", "yellow")
names(cols) <- c("ER+", "Her2+", "TNBC")
annot_colors <- list(Group = cols)


library(viridis)

ER_HER <- up_reg1[up_reg1$adj.P.Val < 0.05 & up_reg1$logFC <= -1, c("logFC", "adj.P.Val")]
ER_HER <- ER_HER[order(ER_HER$logFC),]
write.csv(ER_HER, "/home/clodagh/MASTERS PROJECT/GSE37614/ER_HER_tT_lfc1_down.csv", row.names=T)

HER_TNBC <- up_reg2[up_reg2$adj.P.Val < 0.05 & up_reg2$logFC <= -1, c("logFC", "adj.P.Val")]
HER_TNBC <- HER_TNBC[order(HER_TNBC$logFC),]
#write.csv(HER_TNBC, "/home/clodagh/MASTERS PROJECT/GSE37614/HER_TNBC_tT_lfc1_down.csv", row.names=T)
```

## Top Table

```{r}
tT_df <- tT[rownames(sig_exprs),]
DT::datatable(tT_df, rownames = T, options = list(scrollX = TRUE, pageLength = 7, scroller = TRUE))
```





# TopGO
```{r}
library(topGO)
library(Rgraphviz)
library(org.Hs.eg.db)
library(ggplot2)
library(enrichR)
library(enrichplot)

# DEGs |logFC| over 0.5
list <- tT$adj.P.Val
names(list) <- rownames(top_sig)

# selection is function that returns TRUE/FALSE for p-values<0.05
selection <- function(x) TRUE

allGO2genes <- annFUN.org(whichOnto="BP", feasibleGenes=NULL, mapping="org.Hs.eg.db", ID="symbol")

GOdata <- new("topGOdata", ontology="BP", allGenes=list, annot=annFUN.GO2genes, GO2genes=allGO2genes, geneSel=selection, nodeSize=10)

```
## Perform Enrichment 
```{r}
results.ks <- runTest(GOdata, algorithm="classic", statistic="ks")
goEnrichment <- GenTable(GOdata, KS=results.ks, orderBy="KS", topNodes=20)
goEnrichment$KS <- as.numeric(goEnrichment$KS)
goEnrichment <- goEnrichment[goEnrichment$KS<0.05,]
goEnrichment <- goEnrichment[,c("GO.ID","Term","KS")]
goEnrichment$Term <- gsub(" [a-z]*\\.\\.\\.$", "", goEnrichment$Term)
goEnrichment$Term <- gsub("\\.\\.\\.$", "", goEnrichment$Term)
goEnrichment$Term <- paste(goEnrichment$GO.ID, goEnrichment$Term, sep=", ")
goEnrichment$Term <- factor(goEnrichment$Term, levels=rev(goEnrichment$Term))
#write.table(goEnrichment, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_DEGs_GO_Enrichment", sep = "\t", row.names = F)
```

```{r}
DT::datatable(goEnrichment, options = list(scrollX = TRUE, pageLength = 10, scroller = TRUE))
```

```{r}
rm(mean)
require(ggplot2)
ggplot(goEnrichment, aes(x=Term, y= KS, fill =KS)) +
    stat_summary(geom = "bar", fun = mean, position = "dodge") +
    ylab("Enrichment (-log10(p-value))") +
    ggtitle("GSE37614 GO Biological Process DEGs") +
    scale_y_continuous(breaks = round(seq(0, max(-log10(goEnrichment$KS)), by = 2), 1)) +
    theme_bw(base_size=24) +
    theme(
        legend.position='none',
        legend.background=element_rect(),
        plot.title=element_text(angle=0, size=14, face="bold", vjust=1),
        axis.text.x=element_text(angle=0, size=8, face="bold", hjust=1.10),
        axis.text.y=element_text(angle=0, size=8, face="bold", vjust=0.5),
        axis.title.y = element_blank(),
        axis.title.x =element_text(size=14, face="bold"),
        legend.key=element_blank(),     #removes the border
        legend.key.size=unit(1, "cm"),      #Sets overall area/size of the legend
        legend.text=element_text(size=18),  #Text size
        title=element_text(size=18)) +
        guides(colour=guide_legend(override.aes=list(size=2.5))) +
        
    coord_flip()
```



# Venn Diagram
```{r}
# summarize test results as "up", "down" or "not expressed"
dT <- decideTests(fit2, adjust.method="fdr", p.value=0.05, lfc = 0.5)
vennDiagram(dT, circle.col=palette())

intersected <- rowSums(dT[, c("HER-ER", "HER-TNBC")] != 0L) == 2L
intersect_genes <- names(which(intersected))

```


## Plot of top DEGS between both HER2 and ER+ and HER2 and TNBC
```{r}
intersect_plot <- e[rownames(e) %in% intersect_genes,]
library(RColorBrewer)
library(pheatmap)


### Run pheatmap
pheatmap(intersect_plot[,ord], color =colorRampPalette(rev(brewer.pal(n = 10, name =
"RdYlBu")))(100), cluster_cols =  F, cluster_rows = T, show_rownames=F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "164 overlapping DEGs")

```
## GSEA of 214 overlapping genes

```{r}
library(org.Hs.eg.db)
hs <- org.Hs.eg.db

overlap_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(intersect_plot),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
```


# Fold change over 2
### (ER-Her2)
```{r}
#Top Table of significant genes only
up_reg_fc1 <- top1[top1$adj.P.Val <= 0.05 & abs(top1$logFC) >= 2, ];
nrow(up_reg_fc1)
```
### (Her2-TNBC)
```{r}
#Top Table of significant genes only
up_reg_fc2_ <- top2[top1$adj.P.Val <= 0.05 & abs(top2$logFC) >= 2, ];
nrow(up_reg_fc2_)
```
# combind sig genes fc > 2

```{r}
up_reg_names <- c(rownames(up_reg_fc1), rownames(up_reg_fc2_))
both_sig_exprs <- e[rownames(e) %in% up_reg_names,]
both_sig_exprs <- t(both_sig_exprs)
both_sig_exprs <- scale(both_sig_exprs, center=T, scale = T)
both_sig_exprs <- t(both_sig_exprs)
dim(both_sig_exprs)

```


```{r}
### Set a color palette
heat.colors <- brewer.pal(100, "YlOrRd")

### Run pheatmap
png("/home/clodagh/MASTERS PROJECT/GSE37614/42DEGs_lfc2.png", width = 5, height =6, units = "in", res = 500)
pheatmap(both_sig_exprs[,ord], color = colorRampPalette(rev(brewer.pal(n = 10, name =
"RdYlBu")))(100), cluster_rows = T, show_rownames=T, cluster_cols = F, show_colnames = F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "Top 42 DEGs (|lfc| > 2)")
dev.off()
```
 
# GSEA
```{r}
library(org.Hs.eg.db)
library(DOSE) #Disease ontology; description of gene products with disease perspectvies
library(pathview)
library(clusterProfiler)
library(AnnotationHub)
library(ensembldb)
library(tidyverse)
library(fgsea)
library(stringr) #used to wrap text on dotplot
library(ggplot2)
library(AnnotationDbi)
```



```{r}
all_genes_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(qnorm),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

dim(sig_exprs)
DEGs <-  AnnotationDbi::select(hs, 
       keys = rownames(sig_exprs),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

ego <- enrichGO(gene = DEGs$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                universe = all_genes_entrez$ENTREZID,
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)

cluster_CC_summary <- data.frame(ego)
View(cluster_CC_summary)
write.csv(cluster_CC_summary,'/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_DEGs_MF.csv')
png("/home/clodagh/MASTERS PROJECT/GSE37614/DEGs_MF.png", width= 4.25,
  height = 4.25,
  units = "in",
  res= 1200,)
dotplot(ego, showCategory=10) + ggtitle("DEGs GO (Molecular Function)")  +
  theme_classic()
dev.off()
```






```{r}
#coeff =1
upreg1 <- top1[top1$logFC > 0.5 & top1$adj.P.Val < 0.05, ]
downreg1 <- top1[top1$logFC < -0.5 & top1$adj.P.Val < 0.05, ]

#coeff = 2
upreg2 <- top2[top2$logFC > 0.5 & top2$adj.P.Val < 0.05, ]
downreg2 <- top2[top2$logFC < -0.5 & top2$adj.P.Val < 0.05, ]


up_reg <- c(rownames(upreg1), rownames(upreg2))
up_reg <- unique(up_reg)
down_reg <- c(rownames(downreg1), rownames(downreg2))
down_reg <- unique(down_reg)
```


# get entrez ID from gene symbol 
```{r}

all_genes_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(qnorm),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

DEG_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(top_sigg),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

up_reg_entrez <- AnnotationDbi::select(hs, 
       keys = up_reg,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

down_reg_entrez <-  AnnotationDbi::select(hs, 
       keys = down_reg,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

```

# DEGs overall GSEA (BP)

```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                universe = all_genes_entrez$ENTREZID,
                ont = "BP", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_DEGs_BP.pdf')
dotplot(ego, showCategory=10) + ggtitle("GSE37614 DEGs |logFC| >0.5 GO (Biological Process)")  +
  theme_classic()
dev.off()
```

# DEGs overall GSEA (CC)
```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                universe = all_genes_entrez$ENTREZID,
                ont = "CC", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_CC_summary <- data.frame(ego)
View(cluster_CC_summary)
#can see a llot of terms associated with chromosomal rearrangement
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/DEGS_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("GSE37614 DEGS logFC| >0.5 GO (Cellular Component)")  +
  theme_classic()
dev.off()
```

# DEGs overall GSEA (MF)
```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                universe = all_genes_entrez$ENTREZID,
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_MF_summary <- data.frame(ego)
View(cluster_MF_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/DEGS_MF.pdf')
dotplot(ego, showCategory=10) + ggtitle("GSE37614 DEGs logFC| >0.5 GO (Molecular Function)")  +theme_classic()
dev.off()
```



# DEGs overlap GSEA 
```{r}
library(DOSE)
library(clusterProfiler)
library(AnnotationHub)
library(ensembldb)

# biological process (BP)

ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                universe = all_genes_entrez$ENTREZID,
                OrgDb = org.Hs.eg.db, 
                ont = "BP", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("Overlapping DEGs GO (Biological Process)")  +
  theme_classic()
dev.off()
```

```{r}
# Cellular Compartment
ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                universe = all_genes_entrez$ENTREZID,
                OrgDb = org.Hs.eg.db, 
                ont = "CC", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("Overlapping DEGs GO (Cellular Component)")  +
  theme_classic()
dev.off()
```





```{r}
ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_MF.pdf')
dotplot(ego, showCategory=10) + ggtitle("Overlapping DEGs GO (Molecular Function)")  +
  theme_classic()
dev.off()
```





## GSEA Hallmarks
```{r}
#Prepare file for GSEA input; input ALL genes for testing (including middle ranked (non-significant as removing can reduce power)
#rank genes based on signed pval
top1 <- topTable(fit2, coef=2, adjust="fdr", sort.by="B", number = Inf)
x <- top1
x <- add_rownames(x, var = "symbol") #row to col 
head(x)
x$fcsign <- sign(x$logFC) #returns col of sign (+/- of LFC)
x$logP=-log10(x$P.Value) #take log of all p val
x$metric= x$logP/x$fcsign #signed pval ranking
x <- x[order(-x$metric),] #order in descending order
x <-  na.omit(x) #remove rows with na 
dim(x)
x <- x[!duplicated(x$symbol),] #remove duplicates
```

## REfine BIO tutorial
```{r}
# Attach the library
library(clusterProfiler)

# Package that contains MSigDB gene sets in tidy format
library(msigdbr)

# Zebrafish annotation package we'll use for gene identifier conversion
library(org.Hs.eg.db)

# We will need this so we can use the pipe: %>%
library(magrittr)
```

```{r}
# First let's create a mapped data frame we can join to the differential
# expression stats
dge_mapped_df <- data.frame(
  entrez_id = mapIds(
    # Replace with annotation package for the organism relevant to your data
    org.Hs.eg.db,
    keys = x$symbol,
    # Replace with the type of gene identifiers in your data
    keytype = "SYMBOL",
    # Replace with the type of gene identifiers you would like to map to
    column = "ENTREZID",
    # This will keep only the first mapped value for each symbol
    multiVals = "first"
  )
) %>%
  # If a symbol doesn't map to a Entrez gene identifier,
  # drop that from the data frame
  dplyr::filter(!is.na(entrez_id)) %>%
  # Make an `Ensembl` column to store the rownames
  tibble::rownames_to_column("SYMBOL") %>%
  # Now let's join the rest of the expression data
  dplyr::inner_join(x, by = c("SYMBOL" = "symbol"))

dge_mapped_df

#write.csv(dge_mapped_df, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_dge_mapped_GSEA_table.csv")
```

```{r}
any(duplicated(dge_mapped_df$entrez_id))
```

```{r}
# Let's create a named vector ranked based on the t-statistic values
t_vector <-dge_mapped_df$t
names(t_vector) <- dge_mapped_df$entrez_id

# We need to sort the t-statistic values in descending order here
t_vector <- sort(t_vector, decreasing = TRUE)

# Look at first entries of the ranked t-statistic vector
head(t_vector)
```


# Run GSEA
```{r}
msigdbr_collections()

dr_hallmark_df <- msigdbr(
  species = "Homo sapiens", # H for hallmark
  category = "H"
)


gsea_results <- GSEA(
  geneList = t_vector, # Ordered ranked gene list
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p-value
  seed = TRUE, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  TERM2GENE = dplyr::select(
    dr_hallmark_df,
    gs_name,
    entrez_gene
  )
)
```


```{r}
# We can access the results from our gseaResult object using `@result`
head(gsea_results@result)
gsea_result_df <- data.frame(gsea_results@result)
filter_df <- gsea_result_df[, c("ID", "enrichmentScore", "NES", "p.adjust", "qvalues")]
#write.csv(filter_df, "/home/clodagh/MASTERS PROJECT/GSE37614/DEGs_enriched_hallmarks.csv", row.names=F)
# Visualize 
#Most Positive NES
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(n = 3, order_by = NES)

most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_results,
  geneSetID = "HALLMARK_MITOTIC_SPINDLE",
  title = "HALLMARK_MITOTIC_SPINDLE",
  color.line = "#0d76ff"
)

most_positive_nes_plot
```


# Plot significant results
```{r}
pdf(file="/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_sig_hallmark_pathways.pdf")
#too many terms so just plot significant ones
hallmark <-gsea_result_df[gsea_result_df$p.adjust<0.05,]
write.csv(hallmark, "/home/clodagh/MASTERS PROJECT/GSE37614/HER_TNBC_hallmarks.csv")

png("/home/clodagh/MASTERS PROJECT/GSE37614/DEGs_Hallmarks_HER_TNBC.png")
ggplot(hallmark, aes(reorder(ID, NES), NES)) +
  geom_col(aes(fill=NES < 0)) +
  #scale_fill_manual(name = '', values = setNames(c("#F46D43","#74ADD1"),c("", "")), guide = NULL) +
  scale_fill_manual(values = c("#fa624b","#5e61fc")) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_classic() 
dev.off()
```
# Kegg

# KEGG
```{r}
gsea_results <- gseKEGG(
  geneList = t_vector, # Ordered ranked gene list
  organism = "hsa",
  keyType = "kegg",
  minGSSize = 25, # Minimum gene set size
  maxGSSize = 500, # Maximum gene set set
  pvalueCutoff = 0.05, # p-value cutoff
  eps = 0, # Boundary for calculating the p-value
  seed = TRUE, # Set seed to make results reproducible
  pAdjustMethod = "BH", # Benjamini-Hochberg correction
  by = "fgsea"
  )
gsea_result_KEGG <- data.frame(gsea_results@result)
gsea_result_KEGG <- gsea_result_KEGG %>% arrange(-NES)
keggp <- gsea_result_KEGG[gsea_result_KEGG$p.adjust<0.000000005,]
#write.csv(keggp, "/home/clodagh/MASTERS PROJECT/GSE37614/GSE108565_KEGG_pathways.csv")
ggplot(keggp, aes(reorder(Description, NES), NES)) +
  geom_col(aes(fill=NES < 0)) +
  #scale_fill_manual(name = '', values = setNames(c("#F46D43","#74ADD1"),c("", "")), guide = NULL) +
 scale_fill_manual(values = c("#fa624b","#5e61fc")) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="KEGG NES from GSEA") + 
  theme_classic() 
dev.off()
```