---
title: "GSE37614 batch correction"
author: "Clodagh Murray"
date: "5/19/2022"
output: html_document
---


### Libraries
```{r, message =F}
library(Biobase)
library(ggpubr)
library(data.table)
library(DT)
library(dplyr)
library(ggpubr)
library(ggplot2)
library(limma)
library(RColorBrewer)
library(pheatmap)
```

## Raw Data Staging
### 4 samples were analysed in both batches 1 and 2, creating 4 extra samples in the raw data - update metadata to match raw data 

```{r}
SDRF <- read.delim("/home/clodagh/MASTERS PROJECT/GSE37614/E-GEOD-37614.sdrf_fixed.txt")
rownames(SDRF) <- SDRF$Array.Data.File
DT::datatable(SDRF, options = list(scrollX = TRUE, pageLength = 23, scroller = TRUE))
```

## Count matrix matching metadata?
```{r}
## same as first attempt tbh
gen_raw <- function(path, pattern){
  files = list.files(path, pattern, full.names=T, recursive=T, include.dirs=T)
  mat = as.data.frame(do.call(cbind, lapply(files, function(x) data.table::fread(x, stringsAsFactors=FALSE))))
  ID_REF = mat[,1]
  mat = as.data.frame(mat[,!grepl("ID_REF", colnames(mat))])
  rownames(mat) = ID_REF
  return(mat)
}

raw.dat <- gen_raw("/home/clodagh/MASTERS PROJECT/GSE37614/raw/", "\\.txt$")

## attach sample names to cols
file_names <- list.files("/home/clodagh/MASTERS PROJECT/GSE37614/raw/", "\\.txt$", full.names=T, recursive=T, include.dirs=T)
file_names <- sub(".*/", "", file_names)
sample_names <- sub("\\_.*", "", file_names)
length(sample_names) == ncol(raw.dat)
```

## Append IDs to matrix

```{r}
#  append batch ID's to the sample names for unique colnames. 
SDRF$samp_names <- paste(SDRF$Assay.Name, SDRF$Characteristics..batch., sep=".")
SDRF$samp_names

```
```{r}
# make sure columns match file contents
SDRF <- SDRF[match(file_names, SDRF$Array.Data.File),]
colnames(raw.dat) <- SDRF$samp_names
samples <- SDRF

## the below is correct 
head(raw.dat[1:5, 1:5])
```

## Make Expression Set manually

```{r}
SDRF <- Biobase::AnnotatedDataFrame(SDRF)
rownames(SDRF) <- colnames(raw.dat)
my_set <- Biobase::ExpressionSet(assayData = as.matrix(raw.dat), phenoData = SDRF)
my_set
```

## Raw Data Boxplot
```{r}
SDRF@data$FactorValue..PATIENT.ID.
SDRF@data$Characteristics..class.[23] <- "Her2+"
sml <- as.character(as.numeric(as.factor(SDRF@data$Characteristics..class.))) 
gs <- factor(sml)
groups <- SDRF@data$Characteristics..class. 
orientation <- unique(groups)
orientation <- c(orientation[2], orientation[3], orientation[1])
levels(gs) <- orientation
my_set@phenoData$group <- gs
ord <- order(gs)

palette <- palette(c("#1B9E77", "#7570B3", "pink"))

par(mar=c(7,4,2,1))
title <- "GSE37614 Log2 Raw Signals"
boxplot(log2(exprs(my_set[,ord])), boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", orientation, fill=palette(), bty="n")
```


## Quantile normalization

```{r}
quantile_normalisation <- function(df){
  df_rank <- apply(df,2,rank,ties.method="min")
  df_sorted <- data.frame(apply(df, 2, sort))
  df_mean <- apply(df_sorted, 1, mean)

  index_to_mean <- function(my_index, my_mean){
    return(my_mean[my_index])
  }
df_final <- apply(df_rank, 2, index_to_mean, my_mean=df_mean)
  rownames(df_final) <- rownames(df)
  return(df_final)
}

qnorm <- quantile_normalisation(raw.dat)
```


## Update ExpressionSet Object 

### Boxplot of Qnorm

```{r}
exprs(my_set) <- qnorm

## boxplot
par(mar=c(7,4,2,1))
title <- "GSE37614 Log2 Quantile Normalization"
boxplot(log2(exprs(my_set[,ord])), boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", orientation, fill=palette(), bty="n")
```
## Expression Distribution

```{r}
ex <- exprs(my_set)
par(mar=c(4,4,2,1))
title <- paste ("Normalized Expression Density", sep ="")
limma::plotDensities(log2(ex), group=gs, main=title, legend ="topright")
```

## Probe Filtering

```{r}
# Probe filtering
medians <- rowMedians(ex)

hist(log2(medians + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
```


## Remove Multimapping probes

```{r, message = F}
probes <- rownames(my_set)

library(illuminaHumanv4.db)
```

```{r}
annotation <- AnnotationDbi::select(illuminaHumanv4.db,
                                    keys = probes,
                                    columns = c("SYMBOL", "GENENAME"),
                                    keytype = "PROBEID")
```

```{r}
## 'select()' returned 1:many mapping between keys and columns
annotation <- subset(annotation, !is.na(SYMBOL))

## resolve multi maps. 

ann_grouped <- group_by(annotation, PROBEID)
ann_sum <- dplyr::summarize(ann_grouped, no_of_matches = n_distinct(SYMBOL))

ann_flt <- dplyr::filter(ann_sum, no_of_matches > 1)

remove_id <- (probes %in% ann_flt$PROBEID)

table(remove_id)
```
### Too many genes
*manual filtering*

```{r}
annotations <- read.table("/home/clodagh/MASTERS PROJECT/GSE37614/A-GEOD-10558.adf.txt", header=T, sep="\t")
mart <- biomaRt::useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
attr <- biomaRt::listAttributes(mart)
genbank_map <- biomaRt::getBM(attributes = c("refseq_mrna", "hgnc_symbol", "gene_biotype"), mart = mart, useCache = FALSE)
genbank_map <- genbank_map[!(is.na(genbank_map$refseq_mrna) | genbank_map$refseq_mrna==""),]
genbank_map <- subset(genbank_map, genbank_map$gene_biotype == "protein_coding")

# pop off the version ID
annotations$genbank <- annotations$Reporter.Database.Entry..genbank.

annotations$genbank <- gsub("\\.\\d+$", "", annotations$genbank)
# merge
master <- merge(genbank_map, annotations, by.x="refseq_mrna", by.y="genbank")
```

```{r}
master_sub <- master[,c(4,2)]
master_sub <- master_sub[!(is.na(master_sub$hgnc_symbol) | master_sub$hgnc_symbol==""),]

# remove multi mapping probes
master_sub <- master_sub %>% distinct(hgnc_symbol, .keep_all = T)
dim(master_sub)

```

```{r}
qnorm <- subset(qnorm, rownames(qnorm) %in% master_sub$Reporter.Name)
dim(qnorm)
```

```{r}
master_sub <- master_sub[match(rownames(qnorm), master_sub$Reporter.Name),]
rownames(qnorm) <- master_sub$hgnc_symbol

# qnorm now annotated with gene symbol instead of reporter name
head(qnorm[1:3,1:3])
```

```{r}
medians <- rowMedians(qnorm)

hist(log2(medians + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
```

```{r}
medians <- log2(rowMedians(qnorm) + 1)
qnorm <- as.data.frame(qnorm)
qnorm$medians <- medians

qnorm <- subset(qnorm, qnorm$medians > 7.9)
qnorm <- as.matrix(qnorm[,1:(ncol(qnorm)-1)]) # get rid of last col (medians) 

# plot again
hist(log2(qnorm + 1), 150, col = "cornsilk1", freq = FALSE, 
             main = "Histogram of the median intensities", 
             border = "antiquewhite4", xlab = "Median intensities")
dim(qnorm)
```


# Exploratory Data Analysis

### PCA Cancer Subtypes
*2 groups - largely driven by batch*

```{r}

PCA <- prcomp(t(log2(qnorm)), scale = TRUE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Subtype = my_set$group)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Subtype", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed quantile normalized expression data",
                  subtitle = "Cancer Subtypes",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = FALSE, star.plot = F,
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

### PCA Batch Effects
```{r}
dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Batch = my_set$Characteristics..batch.)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Batch", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed quantile normalized expression data",
                  subtitle = "Batch Effects (Batch 1, Batch 2)",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = TRUE, star.plot = TRUE,
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```


## Pearsons Correlation btween PCs and batch and cancer

```{r}
meta <- my_set@phenoData@data
meta <- meta[,c(4, 5)]
colnames(meta) <- c("Batch", "Cancer_Subtype")
write.table(meta, "metadata.csv", row.names=T, sep="\t")

pca <- PCAtools::pca(log2(qnorm + 1), metadata = meta)

PCAtools::eigencorplot(pca, components = PCAtools::getComponents(pca, 1:10),
    metavars = c('Batch','Cancer_Subtype'),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 1.2, fontCorval = 2, posLab = 'all', rotLabX = 45,
    scale = TRUE,
    main = bquote(PC ~ Pearson ~ r^2 ~ clinical ~ correlates),
    plotRsquared = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'BH',
    signifSymbols = c('****', '***', '**', '*', ''),
    signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1))
```


## Limma Batch Correction
```{r}
groups <- as.factor(groups)
gps <- c("ER", "HER", "TNBC")
design <- model.matrix(~0 + groups)
colnames(design) <- gps
rm_batch <- limma::removeBatchEffect(log2(qnorm + 1), batch = my_set@phenoData@data$Characteristics..batch., design=design)
aw <- arrayWeights(log2(rm_batch), design)
fit <- limma::lmFit(rm_batch, design)

write.table(rm_batch, file = '/home/clodagh/MASTERS PROJECT/GSE37614/GSE37614_counts_matrix.txt', sep ='\t', col.names = T, row.names = T)
```



### Contrast Matrix 
```{r}
# Set up contrasts of interest
cts <- paste(gps, c(tail(gps, -1), head(gps, 1)), sep="-")
cont.matrix <- limma::makeContrasts(contrasts=cts, levels=design)

fit2 <- limma::contrasts.fit(fit, cont.matrix)
fit2 <- limma::eBayes(fit2,0.01)
res <- limma::decideTests(fit2)
summary(res)
```

## PCA Cancer Subtype (Corrected)

```{r}
PCA <- prcomp(t(log2(rm_batch)), scale = TRUE)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Subtype = my_set$group)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Subtype", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed batch corrected data",
                  subtitle = "Cancer Subtypes",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = F, star.plot = T, 
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

## PCA Batch Effect (Corrected)

```{r}
dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Batch = my_set$Characteristics..batch.)


ggscatter(dataGG, x="PC1", y="PC2",
                  color = "Batch", palette = c("dodgerblue4", "darkorange2", "chartreuse", "purple"),
                  title = "PCA plot log-transformed batch corrected data",
                  subtitle = "Batch Effects: Batch1, Batch2",
                  xlab = paste0("PC1, VarExp: ", percentVar[1], "%"),
                  ylab = paste0("PC2, VarExp: ", percentVar[2], "%"),
                  ellipse = T, star.plot = T, 
                  ggtheme = theme_bw()) + 
                  theme(legend.position = "right") + 
                  theme(plot.title = element_text(hjust = 0.5, face = "bold"))
```

# Differentially Expressed Genes
## No coefficient
```{r}
# all genes with a p value < 0.05 and abs(logFC) over 0.5 for any contrast
top1000 <- topTable(fit2, number = Inf)
top_sig <- top1000[top1000$adj.P.Val <= 0.05 & (abs(top1000[,1]) > 0.5 | abs(top1000[,2]) > 0.5), ];
nrow(top_sig)                  
```
## Heatmap 
```{r}
top_sigg <- rm_batch[rownames(rm_batch) %in% rownames(top_sig),]
top_sigg<- t(top_sigg)
top_sigg <- scale(top_sigg, center=T, scale = T)
top_sigg <- t(top_sigg)

annot_col <- data.frame(row.names = colnames(rm_batch), Group = factor(my_set$group))

cols <- c("pink", "blue", "orange")
names(cols) <- c("ER+", "Her2+", "TNBC")
annot_colors <- list(Group = cols)

library(viridis)
```


```{r}
pheatmap::pheatmap(top_sigg[,ord], cluster_cols = F, cluster_rows = T, show_rownames = F,
                   show_colnames = T, annotation_col = annot_col,
                   annotation_colors = annot_colors,
                   col = hcl.colors(100, "Purple-Green",rev=F),
                   cutree_cols = 4, main = "DEGS (logfc >0.5 & p. val < 0.05")
```



### coef = 1 (ER-Her2)
```{r}
#Create the top table for the comparison (coef) we want
top1 <- topTable(fit2, coef=1, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg1 <- top1[top1$adj.P.Val <= 0.05 & abs(top1$logFC) >= 1, ];
nrow(up_reg1)
```
## Volcano plot of ER-HER2
```{r}
plot(top1$logFC, -log10(top1$adj.P.Val), pch="*", xlab="Log2 Fold Change", ylab="-10log (adjusted p-value)", main = "ER vs. Her2 DEGs")
abline(h=-log10(0.05), v=c(-1, 1), col="red", lty=2)
points(up_reg1$logFC, -log10(up_reg1$adj.P.Val), col="red", pch="*")
```

```{r}
library(gplots)
sig_HER_ER <- rm_batch[rownames(rm_batch) %in% rownames(up_reg1),]
sig_HER_ER  <- t(sig_HER_ER)
sig_HER_ER  <- scale(sig_HER_ER , center=T, scale = T)
sig_HER_ER  <- t(sig_HER_ER )
pheatmap::pheatmap(sig_HER_ER[,ord], cluster_cols = F, cluster_rows = T, show_rownames = F,
                   show_colnames = T, annotation_col = annot_col,
                   annotation_colors = annot_colors,
                   col = hcl.colors(100, "Purple-Green",rev=F),
                   cutree_cols = 4, main =  "HER vs ER DEGs")
#heatmap.2(sig_HER_ER, trace="none", scale="row", col="redgreen", cexRow=0.2, cexCol=0.7, main = "HER vs ER DEGS")
```





### coef = 2 (Her2-TNBC)
```{r}
top2 <- topTable(fit2, coef=2, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg2 <- top2[top2$adj.P.Val <= 0.05 & abs(top2$logFC) >= 1, ];
nrow(up_reg2)
```
## Volcano plot of HER2 vs TNBC
```{r}
plot(top2$logFC, -log10(top2$adj.P.Val), pch="*", xlab="Log2 Fold Change", ylab="-10log (adjusted p-value)", main = "Her2 vs TNBC DEGs")
abline(h=-log10(0.05), v=c(-1, 1), col="red", lty=2)
points(up_reg2$logFC, -log10(up_reg2$adj.P.Val), col="red", pch="*")
```

```{r}
library(gplots)
sig_HER_TNBC <- rm_batch[rownames(rm_batch) %in% rownames(up_reg2),]
sig_HER_TNBC  <- t(sig_HER_TNBC)
sig_HER_TNBC  <- scale(sig_HER_TNBC , center=T, scale = T)
sig_HER_TNBC  <- t(sig_HER_TNBC)

pheatmap::pheatmap(sig_HER_TNBC[,ord], cluster_cols = F, cluster_rows = T, show_rownames = F,
                   show_colnames = T, annotation_col = annot_col,
                   annotation_colors = annot_colors,
                   col = hcl.colors(100, "Purple-Green",rev=F),
                   cutree_cols = 4, main =  "HER vs TNBC DEGs")
#heatmap.2(sig_HER_TNBC, trace="none", scale="row", col="redgreen", cexRow=0.2, cexCol=0.7, main = "HER vs TNBC DEGS")
```


### coef = 3 (TNBC-ER)
```{r}
top3 <- topTable(fit2, coef=3, adjust="fdr", sort.by="B", number = Inf)

#Top Table of significant genes only
up_reg3 <-  top3[top3$adj.P.Val <= 0.05, ];
nrow(up_reg3)
```

## Heatmap 

```{r}
up_reg_names <- c(rownames(up_reg1), rownames(up_reg2))
length(up_reg_names)
e <- rm_batch
sig_exprs <- e[rownames(e) %in% up_reg_names,]
sig_exprs <- t(sig_exprs)
sig_exprs <- scale(sig_exprs, center=T, scale = T)
sig_exprs <- t(sig_exprs)
dim(sig_exprs)
annot_col <- data.frame(row.names = colnames(rm_batch), Group = factor(my_set$group))

cols <- c("pink", "blue", "yellow")
names(cols) <- c("ER+", "Her2+", "TNBC")
annot_colors <- list(Group = cols)

library(viridis)
```


```{r}
pheatmap::pheatmap(sig_exprs[,ord], cluster_cols = F, cluster_rows = T, show_rownames = F,
                   show_colnames = T, annotation_col = annot_col,
                   annotation_colors = annot_colors,
                   col = hcl.colors(100, "Purple-Green",rev=F),
                   cutree_cols = 4, main = "DEGs (logFC > 1 & p.val < 0.05)")
```


## Top Table

```{r}
# logFC >= 0.5 & adjusted p-value <= 0.05
write.table(top_sigg, "GSE37614_DEGs_overall.csv", row.names=T, sep="\t")

DT::datatable(sig_exprs, rownames = T, select=c("ER.Her2", "Her2.TNBC", "TNBC.ER","adj.P.Val","P.Value","t"), options = list(scrollX = TRUE, pageLength = 7, scroller = TRUE))
```

# Venn Diagram
```{r}
# summarize test results as "up", "down" or "not expressed"
dT <- decideTests(fit2, adjust.method="fdr", p.value=0.05)
vennDiagram(dT, circle.col=palette())

intersected <- rowSums(dT[, c("ER-HER", "HER-TNBC")] != 0L) == 2L
intersect_genes <- names(which(intersected))

```


## Plot of top DEGS between both HER2 and ER+ and HER2 and TNBC
```{r}
intersect_plot <- e[rownames(e) %in% intersect_genes,]

library(RColorBrewer)
library(pheatmap)
### Set a color palette
heat.colors <- brewer.pal(100, "YlOrRd")


### Run pheatmap
pheatmap(intersect_plot[,ord], color = heat.colors, cluster_cols =  F, cluster_rows = T, show_rownames=F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "410 overlapping DEGs")
```
## GSEA of 401 overlapping genes

```{r}
hs <- org.Hs.eg.db

overlap_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(intersect_plot),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")
```
# DEGs overlap GSEA 

# no biological process (BP)
```{r}
ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "BP", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


#cluster_BP_summary <- data.frame(ego)
#View(cluster_BP_summary)
#pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_BP.pdf')
#dotplot(ego, showCategory=10) + ggtitle("410 Overlapping DEGs GO (Cellular Compartment)")  +
 # theme_classic()
dev.off()
```

# Cellular Compartment
```{r}
ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "CC", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("410 Overlapping DEGs GO (Cellular Compartment)")  +
  theme_classic()
dev.off()
```

```{r}
ego <- enrichGO(gene = overlap_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/overlap_MF.pdf')
dotplot(ego, showCategory=10) + ggtitle("410 Overlapping DEGs GO (Molecular Function)")  +
  theme_classic()
dev.off()
```
# Fold change over 2
### (ER-Her2)
```{r}
#Top Table of significant genes only
up_reg_fc1 <- top1[top1$adj.P.Val <= 0.05 & abs(top1$logFC) >= 2, ];
nrow(up_reg_fc1)
```
### (Her2-TNBC)
```{r}
#Top Table of significant genes only
up_reg_fc2_ <- top2[top1$adj.P.Val <= 0.05 & abs(top2$logFC) >= 2, ];
nrow(up_reg_fc2_)
```
# combind sig genes fc > 2

```{r}
up_reg_names <- c(rownames(up_reg_fc2), rownames(up_reg_fc2_))
both_sig_exprs <- e[rownames(e) %in% up_reg_names,]
both_sig_exprs <- t(both_sig_exprs)
both_sig_exprs <- scale(both_sig_exprs, center=T, scale = T)
both_sig_exprs <- t(both_sig_exprs)
dim(both_sig_exprs)

```


```{r}

### Set a color palette
heat.colors <- brewer.pal(100, "YlOrRd")

### Run pheatmap
pheatmap(both_sig_exprs[,ord], color = heat.colors, cluster_rows = T, show_rownames=T, cluster_cols = F,
annotation= annot_col, border_color=NA, fontsize = 10, scale="row",
     fontsize_row = 10, height=20, main = "Top 27 DEGs (fold change > 2)")
```





# GSEA
```{r}
library(org.Hs.eg.db)
library(DOSE) #Disease ontology; description of gene products with disease perspectvies
library(pathview)
library(clusterProfiler)
library(AnnotationHub)
library(ensembldb)
library(tidyverse)
library(fgsea)
library(stringr) #used to wrap text on dotplot
library(ggplot2)
```


```{r}
#coeff =1
upreg1 <- up_reg1[up_reg1$logFC > 1 & up_reg1$adj.P.Val < 0.05, ]
downreg1 <- up_reg1[up_reg1$logFC < 1 & up_reg1$adj.P.Val < 0.05, ]

#coeff = 2
upreg2 <- up_reg2[up_reg2$logFC > 1 & up_reg2$adj.P.Val < 0.05, ]
downreg2 <- up_reg2[up_reg2$logFC < 1 & up_reg2$adj.P.Val < 0.05, ]


up_reg <- c(rownames(upreg1), rownames(upreg2))
down_reg <- c(rownames(downreg1), rownames(downreg2))
```


```{r}
hs <- org.Hs.eg.db

all_genes_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(rm_batch),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

DEG_entrez <- AnnotationDbi::select(hs, 
       keys = rownames(sig_exprs),
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

up_reg_entrez <- AnnotationDbi::select(hs, 
       keys = up_reg,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

down_reg_entrez <-  AnnotationDbi::select(hs, 
       keys = down_reg,
       columns = c("ENTREZID", "SYMBOL"),
       keytype = "SYMBOL")

```


# DEGs overall GSEA (BP)

```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "BP", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)


cluster_BP_summary <- data.frame(ego)
View(cluster_BP_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/DEGs_BP.pdf')
dotplot(ego, showCategory=10) + ggtitle("DEGs GO (Biological Process)")  +
  theme_classic()
dev.off()
```

# DEGs overall GSEA (CC)
```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "CC", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_CC_summary <- data.frame(ego)
View(cluster_CC_summary)
#can see a llot of terms associated with chromosomal rearrangement
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/DEGS_CC.pdf')
dotplot(ego, showCategory=10) + ggtitle("DEGS GO (Cellular Component)")  +
  theme_classic()
dev.off()
```
# DEGs overall GSEA (MF)
```{r}
ego <- enrichGO(gene = DEG_entrez$ENTREZID, 
                keyType = "ENTREZID",
                OrgDb = org.Hs.eg.db, 
                ont = "MF", #Biological Processes GO term, also done for CC and MF
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
cluster_MF_summary <- data.frame(ego)
View(cluster_MF_summary)
pdf(file='/home/clodagh/MASTERS PROJECT/GSE37614/DEGS_MF.pdf')
dotplot(ego, showCategory=10) + ggtitle("DEGs GO (Molecular Function)")  +
  theme_classic()
dev.off()
```



