---
title: "Research Project RNA-Seq T_vs_N"
author: "Clodagh Murray"
date: "4/15/2022"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Libraries & Functions
```{R, message=F, warning=F}
library(knitr)
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(dplyr)
library(biomaRt)
library(tximport)
library(rhdf5)
library(gplots)
library(DESeq2)
library(GOSemSim)
library(apeglm)
library(org.Hs.eg.db)
library(RColorBrewer)
library(enrichplot)
library(IHW)
library(pheatmap)
library(PCAtools)
library(EnhancedVolcano)
library(pvclust)
library(ComplexHeatmap)
library(circlize)
library(fgsea)
library(DT)
library(clusterProfiler)
library(ggplot2)
library(tidyverse)
library(ggpubr)

get_upregulated <- function(df){

    key <- intersect(rownames(df)[which(df$log2FoldChange>=1)], rownames(df)[which(df$padj<=0.05)])
  results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

get_downregulated <- function(df){

    key <- intersect(rownames(df)[which(df$log2FoldChange<=-1)], rownames(df)[which(df$padj<=0.05)])

    results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

annotate_de_genes <- function(df){

    df$hgnc_symbol <- rownames(df)
    mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
    info <- getBM(attributes=c("hgnc_symbol",
                               "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand",
                               "entrezgene_description"),
                  filters = c("hgnc_symbol"),
                  values = df$hgnc_symbol,
                  mart = mart,
                  useCache=FALSE)

    tmp <- merge(df, info, by="hgnc_symbol")
    tmp$strand <- gsub("-1", "-", tmp$strand)
    tmp$strand <- gsub("1", "+", tmp$strand)
    #tmp$hgnc_symbol <- make.names(tmp$hgnc_symbol, unique = T)
    tmp <- tmp[!grepl("CHR", tmp$chromosome_name),]

    output_col <- c("Gene", "Ensembl ID", "Chromosome", "Start", "Stop", "Strand", "Description", "Log2FC", "P-value", "Adj P-value")
    tmp <- subset(tmp, select=c(hgnc_symbol, ensembl_gene_id_version, chromosome_name, start_position, end_position, strand, entrezgene_description, log2FoldChange, pvalue, padj))
    colnames(tmp) <- output_col

    if(min(tmp$Log2FC) > 0){
        tmp <- tmp[order(-tmp$Log2FC),]
    }else{
        tmp <- tmp[order(tmp$Log2FC),]
    }

    return(tmp)

}

```


# RNA-Seq Analysis
```{r}
samples <- read.csv("/home/clodagh/Downloads/reformat_samples.csv", header=T, row.names = "samples")
samples$ER <- gsub("_negative", "-", samples$ER)
samples$ER <- gsub("_positive", "+", samples$ER)
samples$LVI <- gsub("_negative", "-", samples$LVI)
samples$LVI <- gsub("_positive", "+", samples$LVI)
samples$PR <- gsub("_negative", "-", samples$PR)
samples$PR <- gsub("_positive", "+", samples$PR)
samples$Condition <- gsub("Tumour", "TASC", samples$Condition)
samples$Condition <- gsub("Normal", "TAN", samples$Condition)
samples$Patient <- as.factor(samples$Patient)
samples
```

```{r}
# tx2gene
httr::set_config(httr::config(ssl_verifypeer = FALSE))
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version", "hgnc_symbol"), mart = mart, useCache = FALSE)

#read in salmon
dir <- "/home/clodagh/MASTERS PROJECT/RNA_SEQ//nf-core_results/star_salmon"
files <- file.path(dir, rownames(samples), "quant.sf")
files
names(files) <- rownames(samples)
txi.salmon <- tximport(files, type = "salmon", tx2gene = tx2gene)
dds <- DESeqDataSetFromTximport(txi.salmon, colData = samples, design = ~ Patient + Condition)
dds$Condition <- relevel(dds$Condition, ref = "TAN")
dds <- DESeq(dds)
write.csv(dds, "/home/clodagh/MASTERS PROJECT/RNA_SEQ/dds.csv")
```
# Exploratory Data analysis 

## Data Transformations

### MeanSDPlot
```{R, message=F, warning=F}
library(vsn)
library(hexbin)

nt <- assay(normTransform(dds))
x <- meanSdPlot(nt, ranks=FALSE, plot=FALSE)
x$gg + ggtitle("Log2 + 1 Transformation")

rld <- assay(rlog(dds, blind=FALSE))
rld
y <- meanSdPlot(rld, ranks = FALSE, plot = FALSE)
y$gg + ggtitle("Regularized-Logarithm Transformation")
```


## Sample Heatmap
```{R, fig.width=10, fig.height=6}

subs_samp <- subset(samples, select=-c(Age, Size))

sampleDists <- dist(t(rld))
sampleDistMatrix <- as.matrix(sampleDists)
colnames(sampleDistMatrix) <- rownames(samples)
ann_colors = list(Condition=c(TASC = "forestgreen",  TAN = "gold"),
                  Grade = c(Grade_2 = "dodgerblue4", Grade_3 = "grey3"),
                  Histology = c(Ductal = "magenta3", Lobular = "chartreuse1"),
                  ER = c('ER+' = "black", 'ER-' = "grey67"),
                  LVI = c('LVI-' = "palegreen", 'LVI+' = "palevioletred"),
                  PR = c('PR-' = "royalblue", 'PR+' = "red"))
pheatmap(mat=sampleDistMatrix,
         show_rownames = TRUE,
         cluster_cols = TRUE,
         cluster_rows = TRUE,
         show_colnames = TRUE,
         annotation_col = subs_samp,
         annotation_colors = ann_colors,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colorRampPalette( rev(brewer.pal(9, "Blues")) )(255))
```


## PCA {.tabset} 

### TAN vs. TASC {.tabset}

#### Pairs plot

```{R, message=F, warning=F, fig.width=10, fig.height=8}
p <- pca(rld, metadata = samples)

pairsplot(p, 
          components = getComponents(p, c(1:6)), 
          colby = 'Condition',
          colkey = c('TAN'='royalblue', 'TASC'='red1'), 
          triangle = F,
          plotaxes = T,
          axisLabSize = 8,
          hline = 0,
          hlineType = "dashed",
          hlineCol = "grey10",
          hlineWidth = 0.2,
          vline = 0,
          vlineType = "dashed",
          vlineCol = "grey10",
          vlineWidth = 0.2,
          borderWidth = 0.5,
          title = "PCA pairsplot (components 1:6)\nTAN blue, TASC red",
          titleLabSize = 12)
```

#### PC1 vs PC2

```{R, message=F, warning=F}
biplot(p,
       colby = 'Condition',
       colkey = c('TAN'='royalblue', 'TASC'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC2',
       encircle = T)
```

#### PC1 vs PC3

```{R, message=F, warning=F}

biplot(p,
       x = "PC1", 
       y = "PC3",
       colby = 'Condition',
       colkey = c('TAN'='royalblue', 'TASC'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC3',
       encircle = T)
```

#### PC1 vs PC4

```{R, message=F, warning=F}

biplot(p,
       x = "PC1", 
       y = "PC4",
       colby = 'Condition',
       colkey = c('TAN'='royalblue', 'TASC'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC4',
       encircle = T,
       gridlines.minor = T,
       gridlines.major = T)
```

#### PC1 vs PC5

```{R, message=F, warning=F}

biplot(p,
       x = "PC1", 
       y = "PC5",
       colby = 'Condition',
       colkey = c('TAN'='royalblue', 'TASC'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC5',
       encircle = T)
```

#### PC1 vs PC6

```{R, message=F, warning=F}

biplot(p,
       x = "PC1", 
       y = "PC6",
       colby = 'Condition',
       colkey = c('TAN'='royalblue', 'TASC'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC6',
       encircle = T)
```

#### Loadings

```{R, message=F, warning=F, fig.height=8, fig.width=6}
plt <- plotloadings(p,
             components = getComponents(p, c(1)),
             absolute = F,
    rangeRetain = 0.15,
    labSize = 5.0,
        ylim = c(-0.09,0.125),
    title = 'Loadings plot',
    subtitle = 'PC1',
    caption = 'Top 15% variables',
    shape = 21,
    col = c('royalblue', 'grey70', 'red3'),
    drawConnectors = T, positionConnectors = 'right')

plot(plt)

plt_df <- plt$data
plt_df <- plt_df[order(plt_df$Loading, decreasing = T),]

info <- getBM(attributes=c("hgnc_symbol",
              "entrezgene_description"),
              filters=c("hgnc_symbol"),
              values = plt_df$var,
              mart=mart,
              useCache = F,
              uniqueRows = T)

info <- info[-c(8),]

tmp <- merge(info, plt_df, by.x="hgnc_symbol", by.y="var")
tmp <- tmp[order(tmp$Loading, decreasing = T),]

DT::datatable(tmp, rownames = F)
```

### LVI Status {.tabset}

#### Pairs plot

```{R, message=F, warning=F, fig.width=10, fig.height=8}
pairsplot(p, 
          components = getComponents(p, c(1:6)), 
          colby = 'LVI',
          colkey = c('LVI-'='forestgreen', 'LVI+'='purple'),
          shape = "Condition",
          shapekey = c(16,17),
          triangle = F,
          plotaxes = T,
          axisLabSize = 8,
          hline = 0,
          hlineType = "dashed",
          hlineCol = "grey10",
          hlineWidth = 0.2,
          vline = 0,
          vlineType = "dashed",
          vlineCol = "grey10",
          vlineWidth = 0.2,
          borderWidth = 0.5,
          title = "PCA pairsplot (components 1:6)\nLVI- green, LVI+ purple",
          titleLabSize = 12)
```

#### PC1 vs PC4

```{R, message=F, warning=F}

biplot(p,
       x = "PC4", 
       y = "PC1",
       colby = 'LVI',
       colkey = c('LVI-'='forestgreen', 'LVI+'='purple'), 
       shape = "Condition",
       shapekey = c(16,17),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC4',
       encircle = T)
```

#### PC3 vs PC4

```{R, message=F, warning=F}

biplot(p,
       x = "PC4", 
       y = "PC3",
        colby = 'LVI',
        colkey = c('LVI-'='forestgreen', 'LVI+'='purple'),
          shape = "Condition",
          shapekey = c(16,17),       
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC4 versus PC3',
       encircle = T)
```

#### PC5 vs PC4

```{R, message=F, warning=F}

biplot(p,
       x = "PC4", 
       y = "PC5",
        colby = 'LVI',
        colkey = c('LVI-'='forestgreen', 'LVI+'='purple'), 
                 shape = "Condition",
          shapekey = c(16,17),  
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC4 versus PC5',
       encircle = T)
```

#### Loadings

```{R, message=F, warning=F, fig.height=8, fig.width=6}
plt <- plotloadings(p,
             components = getComponents(p, c(4)),
             absolute = F,
    rangeRetain = 0.15,
    labSize = 5.0,
        ylim = c(-0.09,0.125),
    title = 'Loadings plot',
    subtitle = 'PC4',
    caption = 'Top 15% variables',
    shape = 21,
    col = c('royalblue', 'grey70', 'red3'),
    drawConnectors = T, positionConnectors = 'right')

plot(plt)

plt_df <- plt$data
plt_df <- plt_df[order(plt_df$Loading, decreasing = T),]

info <- getBM(attributes=c("hgnc_symbol",
              "entrezgene_description"),
              filters=c("hgnc_symbol"),
              values = plt_df$var,
              mart=mart,
              useCache = F,
              uniqueRows = T)

info <- info[-c(5),]

tmp <- merge(info, plt_df, by.x="hgnc_symbol", by.y="var")
tmp <- tmp[order(tmp$Loading, decreasing = T),]

DT::datatable(tmp, rownames = F)
```

### PR Status {.tabset}

#### Pairs plot

```{R, message=F, warning=F, fig.width=10, fig.height=8}
pairsplot(p, 
          components = getComponents(p, c(1:6)), 
          colby = 'PR',
          colkey = c('PR-'='darkgoldenrod3', 'PR+'='maroon4'), 
          triangle = F,
          plotaxes = T,
          axisLabSize = 8,
          hline = 0,
          hlineType = "dashed",
          hlineCol = "grey10",
          hlineWidth = 0.2,
          vline = 0,
          vlineType = "dashed",
          vlineCol = "grey10",
          vlineWidth = 0.2,
          borderWidth = 0.5,
          title = "PCA pairsplot (components 1:6)\nPR- darkgold, LVI+ maroon",
          titleLabSize = 12)
```

#### PC1 vs PC2

```{R, message=F, warning=F}

biplot(p,
       x = "PC2", 
       y = "PC1",
        colby = 'PR',
        colkey = c('PR-'='darkgoldenrod3', 'PR+'='maroon4'),
                        shape = "Condition",
          shapekey = c(16,17),  
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC2',
       encircle = T)
```

#### PC2 vs PC3

```{R, message=F, warning=F}

biplot(p,
       x = "PC2", 
       y = "PC3",
        colby = 'PR',
        colkey = c('PR-'='darkgoldenrod3', 'PR+'='maroon4'),
                        shape = "Condition",
          shapekey = c(16,17),  
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC2 versus PC3',
       encircle = T)
```

#### PC2 vs PC4

```{R, message=F, warning=F}

biplot(p,
       x = "PC2", 
       y = "PC4",
        colby = 'PR',
        colkey = c('PR-'='darkgoldenrod3', 'PR+'='maroon4'),
                        shape = "Condition",
          shapekey = c(16,17),  
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC2 versus PC4',
       encircle = T)
```

#### PC2 vs PC5

```{R, message=F, warning=F}

biplot(p,
       x = "PC2", 
       y = "PC5",
        colby = 'PR',
        colkey = c('PR-'='darkgoldenrod3', 'PR+'='maroon4'),
                        shape = "Condition",
          shapekey = c(16,17),  
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       lab = rownames(samples),
       drawConnectors = TRUE,
       title = 'PCA bi-plot',
       subtitle = 'PC2 versus PC5',
       encircle = T)
```

#### Loadings

```{R, message=F, warning=F, fig.height=8, fig.width=6}
plt <- plotloadings(p,
             components = getComponents(p, c(2)),
             absolute = F,
    rangeRetain = 0.15,
    labSize = 5.0,
        ylim = c(-0.09,0.125),
    title = 'Loadings plot',
    subtitle = 'PC2',
    caption = 'Top 15% variables',
    shape = 21,
    col = c('royalblue', 'grey70', 'red3'),
    drawConnectors = T, positionConnectors = 'right')

plot(plt)

plt_df <- plt$data
plt_df <- plt_df[order(plt_df$Loading, decreasing = T),]

info <- getBM(attributes=c("hgnc_symbol",
              "entrezgene_description"),
              filters=c("hgnc_symbol"),
              values = plt_df$var,
              mart=mart,
              useCache = F,
              uniqueRows = T)


tmp <- merge(info, plt_df, by.x="hgnc_symbol", by.y="var")
tmp <- tmp[order(tmp$Loading, decreasing = T),]

DT::datatable(tmp, rownames = F)
```

## Clinical Correlations

```{R, message=F, warning=F, fig.width=10}
  eigencorplot(p,
    components = getComponents(p, 1:10),
    metavars = colnames(samples),
    col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
    cexCorval = 0.7,
    colCorval = 'black',
    fontCorval = 2,
    posLab = 'bottomleft',
    rotLabX = 45,
    posColKey = 'top',
    cexLabColKey = 1.5,
    scale = TRUE,
    corFUN = 'pearson',
    corUSE = 'pairwise.complete.obs',
    corMultipleTestCorrection = 'none',
    main = 'PC1-10 clinical correlations',
    colFrame = 'white',
    plotRsquared = TRUE)
```

# Differentially Expressed Genes TAN vs. TASC {.tabset}

```{R, message=F, warning=F}
res <- results(dds, filterFun = ihw, alpha = 0.05, c("Condition", "TASC", "TAN"))
lfc <- lfcShrink(dds = dds, res = res, coef = 13, type = "apeglm")
lfc_df <- as.data.frame(lfc)

up <- get_upregulated(lfc_df)
down <- get_downregulated(lfc_df)

up <- annotate_de_genes(up)
down <- annotate_de_genes(down)
```

## Up Regulated TASC

```{R, message=F, warning=F}
DT::datatable(up, rownames = FALSE, options=list(scrollX=T))
```

## Down Regulated TASC

```{R, message=F, warning=F}
DT::datatable(down, rownames = FALSE, options=list(scrollX=T))
```

## MA Plot

Below demonstrates the effect of `apeglm`, the shrinkage estmator designed by the authors of `DESeq2`. Per the `apeglm` paper: 

> When the counts of sequenced reads are small or have a high coefficient of variation in one or a subset of the conditions, the estimated LFCs will have high variance, leading to some large estimated LFCs, which do not represent true large differences in expression ... Here, we propose the use of a heavy-tailed Cauchy prior distribution for effect sizes, which avoids the use of filter thresholds (removing lowly expressed genes -- problem with this is where do we set the cutoff?) or pseudocounts. The proposed method, Approximate Posterior Estimation for generalized linear model, `apeglm`, has lower bias than previously proposed shrinkage estimators, while still reducing variance for those genes with little information for statistical inference.

Essentially, the formula shrinks the effect size (Log2FoldChange) of genes with low counts and high variability (recall the WT1 boxplot, given below) while preserving the true large effect sizes in the biological contrast of interest (TASC vs. TAN). 

This shrinkage is evident in the MAplots given below, which plot each genes normalised expression and log2foldchange value. 

```{R, fig.height=5, fig.width=5}
plotCounts(dds, gene="WT1", intgroup="Condition")
```


```{R, message=F, warning=F, fig.width=8, fig.heigth=10}
maplot <- plotMA(res, returnData=T)
maplot$isDE <- ifelse(maplot$lfc >= 1 | maplot$lfc <= -1, TRUE, FALSE)
#maplot$isDE <- ifelse(maplot$lfc <= -1, TRUE, FALSE)
plotMA(maplot, ylim=c(-3,3), main = "TASC vs. TAN DEGs LFC > |1| & alpha = 0.05 \n no shrinkage", alpha = 0.05)
```

```{R, message=F, warning=F, fig.width=8, fig.heigth=10}
maplot <- plotMA(lfc, returnData=T)
maplot$isDE <- ifelse(maplot$lfc >= 1 | maplot$lfc <= -1, TRUE, FALSE)
#maplot$isDE <- ifelse(maplot$lfc <= -1, TRUE, FALSE)
plotMA(maplot, ylim=c(-3,3), main = "TASC vs. TAN DEGs LFC > |1| & alpha = 0.05", alpha = 0.05)
```

## Volcano Plot

```{R, message=F, warning=F, fig.width=10, fig.height=8}
  EnhancedVolcano(lfc,
    lab = rownames(lfc),
    labSize = 0,
    pointSize = 1.2,
    FCcutoff = 1, 
    pCutoff = 0.05,
    title = "TASC vs TAN LFC > |1| & padj > 0.05",
    subtitle = "",
    legendPosition = 'none',
    pCutoffCol = 'padj',
    x = 'log2FoldChange',
    y = 'padj',
    ylim = c(0, 10.5))


```

## Heatmap (A)

Global Expression pattern of DEGS with a LFC cutoff of 1 (i.e at least a fold change of 2), and padj < 0.05 in TASC vs. TAN. 328 genes, thus gene labels have been omitted from the rows.  

```{R, message=F, warning=F, fig.height=12, fig.width=10}
master <- rbind(up,down)

mat <- subset(rld, rownames(rld) %in% master$Gene)

mat <- t(mat)
mat <- scale(mat, center=T, scale=T)
mat <- t(mat)

# drop 4722
#mat <- mat[,-23]

# set up annotation dataframe
ann <- data.frame(Cell_Type = c(samples$Condition))

#ann <- data.frame(Cell_Type = c(samples$Condition),
 #                 samples = c(rownames(samples)))

#ann <- subset(ann, ann$samples != 4722)
#ann <- subset(ann, select=-c(samples))

# set up heatmap column annotation
ha_col = HeatmapAnnotation(df = ann,
                           col = list(Cell_Type = c("TAN" =  "red",
                                                    "TASC" = "blue")),
                           annotation_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                                          labels_gp = gpar(fontsize = 12)),
                          annotation_name_side = "right")

# set up heatmap object
hm1 <- Heatmap(mat,
              col= colorRamp2(c(-2,-1,0,2,4),c("green", "green2", "black", "red2", "red")),
              heatmap_legend_param=list(at=c(-2,-1,0,2,4),color_bar="continuous",
                                        legend_direction="vertical", legend_width=unit(5,"cm"),
                                        title_position="topcenter", title_gp=gpar(fontsize=10, fontface="bold")),
              name = "Z-score",

              #Row annotation configurations
              cluster_rows=T,
              show_row_dend=T,
              row_title_side="right",
              row_title_gp=gpar(fontsize=4),
              show_row_names=FALSE,
              row_names_side="right",

              #Column annotation configuratiions
              cluster_columns=T,
              show_column_dend=T,
              column_title="TASC vs. TAN DEG LFC > |1|",
              #column_title_side="top",
              column_title_gp=gpar(fontsize=15, fontface="bold"),
              show_column_names = T,
              column_names_gp = gpar(fontsize = 12, fontface="bold"),

              #Dendrogram configurations: columns
              clustering_distance_columns="euclidean",
              clustering_method_columns="complete",
              column_dend_height=unit(2,"cm"),

              #Dendrogram configurations: rows
              clustering_distance_rows="euclidean",
              clustering_method_rows="complete",
              row_dend_width=unit(2,"cm"),
              row_dend_side = "left",
              row_dend_reorder = TRUE,

              #Splits
              border=T,
              row_km = 1,
              column_km = 1,

              #plot params
              #width = unit(5, "inch"),
              #height = unit(4, "inch"),
              #height = unit(0.4, "cm")*nrow(mat),

              #Annotations
              top_annotation = ha_col)

# plot heatmap
draw(hm1, annotation_legend_side = "right", heatmap_legend_side="right")
```

## Heatmap (B)

Global expression pattern of DEGs with a LFC of 2 (a fold change of 4 vs. normal) and padj < 0.05. 

```{R, message=F, warning=F, fig.width=10, fig.height=12}
# use strict cutoff for heatmaps (LFC 2)
get_upregulated_hm <- function(df){

    key <- intersect(rownames(df)[which(df$log2FoldChange>=2)], rownames(df)[which(df$padj<=0.05)])

  results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

get_downregulated_hm <- function(df){

    key <- intersect(rownames(df)[which(df$log2FoldChange<=-2)], rownames(df)[which(df$padj<=0.05)])

    results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}

up_hm <- get_upregulated_hm(lfc_df)
down_hm <- get_downregulated_hm(lfc_df)

up_hm <- annotate_de_genes(up_hm)
down_hm <- annotate_de_genes(down_hm)

master <- rbind(up_hm,down_hm)

mat <- subset(rld, rownames(rld) %in% master$Gene)

mat <- t(mat)
mat <- scale(mat, center=T, scale=T)
mat <- t(mat)

# drop 4722
#mat <- mat[,-23]

# set up annotation dataframe
ann <- data.frame(Cell_Type = c(samples$Condition))

#ann <- data.frame(Cell_Type = c(samples$Condition),
 #                 samples = c(rownames(samples)))

#ann <- subset(ann, ann$samples != 4722)
#ann <- subset(ann, select=-c(samples))

# set up heatmap column annotation
ha_col = HeatmapAnnotation(df = ann,
                           col = list(Cell_Type = c("TAN" =  "red",
                                                    "TASC" = "blue")),
                           annotation_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                                          labels_gp = gpar(fontsize = 12)),
                          annotation_name_side = "right")

# set up heatmap object
hm1 <- Heatmap(mat,
              col= colorRamp2(c(-2,-1,0,2,4),c("green", "green2", "black", "red2", "red")),
              heatmap_legend_param=list(at=c(-2,-1,0,2,4),color_bar="continuous",
                                        legend_direction="vertical", legend_width=unit(5,"cm"),
                                        title_position="topcenter", title_gp=gpar(fontsize=10, fontface="bold")),
              name = "Z-score",

              #Row annotation configurations
              cluster_rows=T,
              show_row_dend=T,
              row_title_side="right",
              row_names_gp = gpar(fontsize=8, fontface='bold'),
              show_row_names=TRUE,
              row_names_side="right",

              #Column annotation configuratiions
              cluster_columns=T,
              show_column_dend=T,
              column_title="TASC vs. TAN DEG LFC > |2|",
              #column_title_side="top",
              column_title_gp=gpar(fontsize=15, fontface="bold"),
              show_column_names = T,
              column_names_gp = gpar(fontsize = 12, fontface="bold"),

              #Dendrogram configurations: columns
              clustering_distance_columns="euclidean",
              clustering_method_columns="complete",
              column_dend_height=unit(2,"cm"),

              #Dendrogram configurations: rows
              clustering_distance_rows="euclidean",
              clustering_method_rows="complete",
              row_dend_width=unit(2,"cm"),
              row_dend_side = "left",
              row_dend_reorder = TRUE,

              #Splits
              border=T,
              row_km = 1,
              column_km = 1,

              #plot params
              #width = unit(5, "inch"),
              #height = unit(4, "inch"),
              #height = unit(0.35, "cm")*nrow(mat),

              #Annotations
              top_annotation = ha_col)

# plot heatmap
draw(hm1, annotation_legend_side = "right", heatmap_legend_side="right")

```

## Heatmap (C)

Global expression pattern of DEGs with a LFC of 2 (a fold change of 4 vs. normal) and padj < 0.05, with wet lab genes validated by Domhnall highlighted. 

Please note that `WT1` is no longer considered differentially expressed after applying more robust filtering methods to the analysis.

```{R, message=F, warning=F, fig.height=12, fig.width=10}
#master <- rbind(up,down)

#mat <- subset(rld, rownames(rld) %in% master$Gene)

#mat <- t(mat)
#mat <- scale(mat, center=T, scale=T)
#mat <- t(mat)


myRows <- c("SLC7A2", "GPR37", "ADGRF5", "PCDH10")

# Set stylings for row names and make our selected rows unique
row_idx <- which(rownames(mat) %in% myRows)
fontsizes <- rep(10, nrow(mat))
fontsizes[row_idx] <- 12
fontcolors <- rep('black', nrow(mat))
fontcolors[row_idx] <- 'red'
fontfaces <- rep('plain',nrow(mat))
fontfaces[row_idx] <- 'bold'

# Create text annotation object for displaying row names
rowAnno <- rowAnnotation(rows = anno_text(rownames(mat), gp = gpar(fontsize = fontsizes, fontface = fontfaces, col = fontcolors)))

# Create our own row dendrogram (ComplexHeatmap orders rows by mean by default)
dend <- reorder(as.dendrogram(hclust(dist(mat))), -rowMeans(mat), agglo.FUN = mean)

# Find rows in dendrogram
dend_idx <- which(order.dendrogram(dend) %in% which(rownames(mat) %in% myRows))

# Find bottom and top of each row on heatmap (x and y axes go from 0 to 1)
btm <- 1 - (dend_idx / nrow(mat))
top <- btm + (1/nrow(mat))

# set up annotation dataframe
ann <- data.frame(Cell_Type = c(samples$Condition))

# set up heatmap column annotation
ha_col = HeatmapAnnotation(df = ann,
                           col = list(Cell_Type = c("TAN" =  "red",
                                                    "TASC" = "blue")),
                           annotation_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                                          labels_gp = gpar(fontsize = 12)),
                          annotation_name_side = "left")

Heatmap(mat, name = "Zscore", cluster_rows = dend, right_annotation = rowAnno, show_row_names = FALSE, top_annotation = ha_col,
        col= colorRamp2(c(-2,-1,0,2,4),c("green", "green2", "black", "red2", "red")),
              heatmap_legend_param=list(at=c(-2,-1,0,2,4),color_bar="continuous",
                                        legend_direction="vertical", legend_width=unit(5,"cm"),
                                        title_position="topcenter", title_gp=gpar(fontsize=10, fontface="bold")), show_column_names = T,
              column_names_gp = gpar(fontsize = 12, fontface="bold"), border = T, row_dend_width=unit(2,"cm"), column_dend_height = unit(2,"cm"))

box_col <- 'black'
box_width <- 2
decorate_heatmap_body("Zscore", { for (i in 1:length(myRows)) {
  grid.lines(c(0, 1), c(top[i],top[i]), gp = gpar(lty = 1, lwd = box_width, col = box_col))
  grid.lines(c(0, 1), c(btm[i],btm[i]), gp = gpar(lty = 1, lwd = box_width, col = box_col))
  grid.lines(c(0, 0), c(btm[i],top[i]), gp = gpar(lty = 1, lwd = box_width, col = box_col))
  grid.lines(c(1, 1), c(btm[i],top[i]), gp = gpar(lty = 1, lwd = box_width, col = box_col))
}
})
```

# Pathway Analysis {.tabset}

```{R, message=F, warning=F}
master_lfc1 <- rbind(up,down)
# want to create 'background' gene set entrez id + LFC values for all genes
info <- getBM(attributes = c("hgnc_symbol", "entrezgene_id"),
              mart=mart, filters = "hgnc_symbol", values = rownames(lfc_df))

lfc_df$Gene <- rownames(lfc_df)
tmp <- merge(info, lfc_df, by.x="hgnc_symbol", by.y="Gene")

background <- tmp$log2FoldChange
names(background) <- tmp$hgnc_symbol
background <- sort(background, decreasing = TRUE)

# make sure you have all DEGs here (LFC > 1 and pval cutoff i.e maser_lfc1 genes)
degtmp <- subset(tmp, tmp$hgnc_symbol %in% master_lfc1$Gene)
deg <- degtmp$entrezgene_id
deg <- na.omit(deg)

deg.df <- bitr(deg, fromType = "ENTREZID",
                toType = c("ENSEMBL", "SYMBOL"),
                OrgDb = org.Hs.eg.db)
```

## Hallmarks {.tabset}

```{R, message=F, warning=F}
hmarks <- read.gmt("/home/clodagh/MASTERS PROJECT/RNA_SEQ/h.all.v7.4.symbols.gmt")

egmt <- GSEA(background,
             TERM2GENE=hmarks,
             pvalueCutoff = 0.1, 
             pAdjustMethod = 'BH',
             minGSSize = 1, 
             maxGSSize = 1000,
             by='fgsea',
             seed=11384191)

egmt_df <- egmt@result

egmt_subs <- subset(egmt_df, select=c(Description, enrichmentScore, NES, pvalue, p.adjust, core_enrichment ))

DT::datatable(egmt_subs, rownames = FALSE, options=list(scrollX=T))
```

### Dot Plot

```{R, message=F, warning=F, fig.height=5, fig.width=7}
dotplot(egmt, title = "GSEA HALLMARKS", x="NES")
```

### Enrichment plots

```{R, message=F, warning=F, fig.width=10, fig.height=5}
for(i in 1:nrow(egmt_df)){
  p <- gseaplot2(egmt, geneSetID = i, title = egmt_df$Description[[i]], pvalue_table = T)
  plot(p)
}
```

## GO Biological Processes {.tabset}

```{R,message=F, warning=F}
gobp <- read.gmt("/home/clodagh/MASTERS PROJECT/RNA_SEQ/c5.go.bp.v7.4.symbols.gmt")

egmt <- GSEA(background,
             TERM2GENE=gobp,
             pvalueCutoff = 0.1, 
             pAdjustMethod = 'BH',
             minGSSize = 1, 
             maxGSSize = 1000,
             by='fgsea',
             seed=11384191)

egmt_df <- egmt@result
#egmt_df$Description <- gsub("GOBP_", "", egmt_df$Description)
#egmt_df$ID <- gsub("GOBP_", "", egmt_df$ID)

egmt_subs <- subset(egmt_df, select=c(Description, enrichmentScore, NES, pvalue, p.adjust, core_enrichment ))

DT::datatable(egmt_subs, rownames = FALSE, options=list(scrollX=T))
```

### Dot Plot

```{R, message=F, warning=F, fig.height=10, fig.width=12}
dotplot(egmt, title = "GO BIOLOGICAL PROCESSES", x="NES", showCategory=29)
```

### Enrichment Plots

```{R, message=F, warning=F, fig.width=10, fig.height=5}
for(i in 1:nrow(egmt_df)){
  p <- gseaplot2(egmt, geneSetID = i, title = egmt_df$Description[[i]], pvalue_table = T)
  plot(p)
}
```

### Tree Plot

```{R, warning=F, message=F, fig.width=14, fig.height=10}
library(ggnewscale)
d <- godata('org.Hs.eg.db', ont="BP")
ego2 <- pairwise_termsim(egmt, method = "JC", semData = d)
tree <- treeplot(ego2, showCategory = 29, color = "pvalue", fontsize = 0, offset = 100, nCluster=7)
plot(tree)
```

### Cluster 1 

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster1 <- subset(tree$data$label, tree$data$group == "cluster_1")

cluster1 <- na.omit(cluster1)

p <- cnetplot(egmt, showCategory = cluster1, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 2

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster2 <- subset(tree$data$label, tree$data$group == "cluster_2")

cluster2 <- na.omit(cluster2)

p <- cnetplot(egmt, showCategory = cluster2, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 3

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster3 <- subset(tree$data$label, tree$data$group == "cluster_3")

cluster3 <- na.omit(cluster3)

p <- cnetplot(egmt, showCategory = cluster3, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 4

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster4 <- subset(tree$data$label, tree$data$group == "cluster_4")

cluster4 <- na.omit(cluster4)

p <- cnetplot(egmt, showCategory = cluster4, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 5

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster5 <- subset(tree$data$label, tree$data$group == "cluster_5")

cluster5 <- na.omit(cluster5)

p <- cnetplot(egmt, showCategory = cluster5, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

min.value <- floor( min(p$data$color, na.rm = TRUE) )
max.value <- ceiling( max(p$data$color, na.rm = TRUE) )
p + scale_color_gradientn(name = "fold change",
                       colours = c("blue", "grey", "red"),
                       limits= c(min.value, max.value), 
                       breaks=c(min.value , 0, max.value))
```

### Cluster 6

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster6 <- subset(tree$data$label, tree$data$group == "cluster_6")

cluster6 <- na.omit(cluster6)

p <- cnetplot(egmt, showCategory = cluster6, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("grey", "red"), name = "Log2FoldChange")
```

### Cluster 7

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster7 <- subset(tree$data$label, tree$data$group == "cluster_7")

cluster7 <- na.omit(cluster7)

p <- cnetplot(egmt, showCategory = cluster7, foldChange = background, circular=T, colorEdge=T, node_label = 'none', cex_label_category = 1, cex_label_gene=0.5, cex_gene = 0.75)

p + scale_color_gradientn(colours = c("grey", "red"), name = "Log2FoldChange")
```

## GO Cellular Component {.tabset}

```{R,message=F, warning=F}
gocc <- read.gmt("/home/clodagh/MASTERS PROJECT/RNA_SEQ/c5.go.cc.v7.4.symbols.gmt")

egmt <- GSEA(background,
             TERM2GENE=gocc,
             pvalueCutoff = 0.1, 
             pAdjustMethod = 'BH',
             minGSSize = 1, 
             maxGSSize = 1000,
             by='fgsea',
             seed=11384191)

egmt_df <- egmt@result

egmt_df$ID <- gsub("GOCC_", "", egmt_df$ID)

egmt_subs <- subset(egmt_df, select=c(Description, enrichmentScore, NES, pvalue, p.adjust, core_enrichment ))

DT::datatable(egmt_subs, rownames = FALSE, options=list(scrollX=T))
```

### Dot Plot

```{R, message=F, warning=F, fig.height=10, fig.width=10}
dotplot(egmt, title = "GO CELLULAR COMPONENT", x="NES", showCategory=24)
```

### Enrichment Plots

```{R, message=F, warning=F, fig.width=10, fig.height=5}
for(i in 1:nrow(egmt_df)){
  p <- gseaplot2(egmt, geneSetID = i, title = egmt_df$Description[[i]], pvalue_table = T)
  plot(p)
}
```

### Tree Plot

```{R, warning=F, message=F, fig.width=14, fig.height=10}
#d <- godata('org.Hs.eg.db', ont="BP")
ego2 <- pairwise_termsim(egmt, method = "JC", semData = d)
tree <- treeplot(ego2, showCategory = 24, color = "pvalue", fontsize = 0, offset = 100, nCluster=5)
plot(tree)
```

### Cluster 1

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster1 <- subset(tree$data$label, tree$data$group == "cluster_1")

cluster1 <- na.omit(cluster1)

p <- cnetplot(egmt, showCategory = cluster1, foldChange = background, circular=T, colorEdge=T, node_label = 'none', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 2

Anchoring junction has been dropped due to a huge number of non-de genes, crowding the plot. 

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster2 <- subset(tree$data$label, tree$data$group == "cluster_2")

cluster2 <- na.omit(cluster2)
cluster2 <- cluster2[-2]

p <- cnetplot(egmt, showCategory = cluster2, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("grey", "red"), name = "Log2FoldChange")
```

### Cluster 3

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster3 <- subset(tree$data$label, tree$data$group == "cluster_3")

cluster3 <- na.omit(cluster3)

p <- cnetplot(egmt, showCategory = cluster3, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

### Cluster 4

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster4 <- subset(tree$data$label, tree$data$group == "cluster_4")

cluster4 <- na.omit(cluster4)

p <- cnetplot(egmt, showCategory = cluster4, foldChange = background, circular=T, colorEdge=T, node_label = 'none', cex_label_category = 1, cex_label_gene=0.75)

min.value <- floor( min(p$data$color, na.rm = TRUE) )
max.value <- ceiling( max(p$data$color, na.rm = TRUE) )
p + scale_color_gradientn(name = "fold change",
                       colours = c("blue", "grey", "red"),
                       limits= c(min.value, max.value), 
                       breaks=c(min.value , 0, max.value))
```

### Cluster 5

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster5 <- subset(tree$data$label, tree$data$group == "cluster_5")

cluster5 <- na.omit(cluster5)

p <- cnetplot(egmt, showCategory = cluster5, foldChange = background, circular=T, colorEdge=T, node_label = 'gene', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```

## GO Molecular Function {.tabset}

```{R,message=F, warning=F}
gomf <- read.gmt("/home/clodagh/MASTERS PROJECT/RNA_SEQ/c5.go.mf.v7.4.symbols.gmt")

egmt <- GSEA(background,
             TERM2GENE=gomf,
             pvalueCutoff = 0.1, 
             pAdjustMethod = 'BH',
             minGSSize = 1, 
             maxGSSize = 1000,
             by='fgsea',
             seed=11384191)

egmt_df <- egmt@result

egmt_subs <- subset(egmt_df, select=c(Description, enrichmentScore, NES, pvalue, p.adjust, core_enrichment ))

DT::datatable(egmt_subs, rownames = FALSE, options=list(scrollX=T))
```

### Dot Plot

```{R, message=F, warning=F, fig.height=6, fig.width=8}
dotplot(egmt, title = "GO MOLECULAR FUNCTION", x="NES", showCategory=8)
```

### Enrichment Plots

```{R, message=F, warning=F, fig.width=10, fig.height=5}
for(i in 1:nrow(egmt_df)){
  p <- gseaplot2(egmt, geneSetID = i, title = egmt_df$Description[[i]], pvalue_table = T)
  plot(p)
}
```

### Tree Plot

```{R, warning=F, message=F, fig.width=14, fig.height=10}
#d <- godata('org.Hs.eg.db', ont="BP")
ego2 <- pairwise_termsim(egmt, method = "JC", semData = d)
tree <- treeplot(ego2, showCategory = 8, color = "pvalue", fontsize = 0, offset = 100, nCluster=3)
plot(tree)
```

### Cluster 1

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster1 <- subset(tree$data$label, tree$data$group == "cluster_1")

cluster1 <- na.omit(cluster1)
cluster1 <- cluster1[-1]
p <- cnetplot(egmt, showCategory = cluster1, foldChange = background, circular=T, colorEdge=T, node_label = 'all', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("grey", "red"), name = "Log2FoldChange")
```


### Cluster 2

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster2 <- subset(tree$data$label, tree$data$group == "cluster_2")

cluster2 <- na.omit(cluster2)

p <- cnetplot(egmt, showCategory = cluster2, foldChange = background, circular=T, colorEdge=T, node_label = 'all', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("blue", "grey"), name = "Log2FoldChange")
```


### Cluster 3

```{R, message=F, warning=F, fig.height=10, fig.width=15}
cluster3 <- subset(tree$data$label, tree$data$group == "cluster_3")

cluster3 <- na.omit(cluster3)

p <- cnetplot(egmt, showCategory = cluster3, foldChange = background, circular=T, colorEdge=T, node_label = 'all', cex_label_category = 1, cex_label_gene=0.75)

p + scale_color_gradientn(colours = c("grey", "red"), name = "Log2FoldChange")
```


### Hub gene expression boxplots

```{r}
# Extract normalized values 
wpn_vsd <- getVarianceStabilizedData(dds)

## sample traits
grade <- factor(samples$Grade)
grades <- make.names(c("Grade 2", "Grade 3"))
levels(grade) <- grades

gs <- factor(samples$Condition)
groups <- make.names(c("Tumour","Normal"))
levels(gs) <- groups
colnames(wpn_vsd)  <- paste(gs, "-", samples$Patient)
ord <- order(gs)  # order samples by group
```


# Hubs 
```{r}
library(ggpubr)
library(reshape)
normal_idx <- c(2,4,6,8,10,12,14,16,18,10,22,24)
  
Res_hubs <- c("IFI44L","DDX60","DHX58","IFIT3","OAS3","MX2","LAP3","EPSTI1","SAMD9L","GCH1","UBE2L6","IFIH1","RTP4","IL15RA","IFIT2","USP18","SLC39A14","NFKBIA","XAF1","IFI44")
HER2_hubs <- c("PFN1","DDA1","ARHGEF17","TNS1","OXTR","LYPD6B","SHROOM3","CSNK1D","CAP1","TPM1","DYSF","YIF1B","CTPS1","CHN1","UCK2","PCED1B","CALM2","ITGA3","MICAL2")
```

# Subtype Hubs
```{r}
#png("/home/clodagh/MASTERS PROJECT/GSE37614/RNA_seq_hubs.png", width = 6, height=6, res= 500, units = "in")
#compare_means(normalized_counts ~ group, data = melted_norm_counts, 
#              group.by = "gene")

#melting dataset for visualization
melted_norm_counts <-data.frame(melt(wpn_vsd[HER2_hubs,]))
colnames(melted_norm_counts) <- c("gene", "samplename", "normalized_counts")
melted_norm_counts$group <- ifelse(melted_norm_counts$samplename %in% colnames(assay(vsd))[normal_idx], "Normal", "Tumor")

png("/home/clodagh/MASTERS PROJECT/GSE37614/RNA_seq_hubs.png", width = 7, height=6, res= 500, units = "in")

#plot
ggplot(melted_norm_counts, aes(x = as.factor(gene), y = normalized_counts)) +
geom_boxplot(aes(fill = group), position = position_dodge(0.9)) +
facet_wrap(~ gene, scales = "free") +
scale_fill_manual(values = c("#09E359", "#E31009")) + 
ylab("Normalized Expression") +
labs(fill = element_blank()) +
theme_bw() +
theme(
  axis.title.x = element_blank(),
)
#dev.off()
```

# Resistance Hubs
```{r}
#melting dataset for visualization
melted_norm_counts <-data.frame(melt(wpn_vsd[Res_hubs,]))
colnames(melted_norm_counts) <- c("gene", "samplename", "normalized_counts")
melted_norm_counts$group <- ifelse(melted_norm_counts$samplename %in% colnames(assay(vsd))[normal_idx], "Normal", "Tumor")

#plot
ggplot(melted_norm_counts, aes(x = as.factor(gene), y = normalized_counts)) +
geom_boxplot(aes(fill = group), position = position_dodge(0.9)) +
facet_wrap(~ gene, scales = "free") +
scale_fill_manual(values = c("#09E359", "#E31009")) + 
ylab("Normalized Expression") +
labs(fill = element_blank()) +
theme_bw() +
theme(
  axis.title.x = element_blank(),
)
#dev.off()
```
